---
layout: tutorial_page
permalink: /EPI_2021_Module3_lab
title: EPI 2021 Module 3 Lab
header1: Workshop Pages for Students
header2: Epigenomic Analysis 2021
image: /site_images/CBW_Epigenome-data_icon.jpg
home: https://bioinformaticsdotca.github.io/EPI_2021
description: Introduction to ChIP-seq and analysis
author: Martin Hirst and Edmund Su

---

# Introduction to ChIP-seq and analysis

*by Martin Hirst and Edmund Su*

# Table of contents:

<font size="5">[Common tools of the trade](#Common-tools-of-the-trade)</font>

<font size="5">[Module 3. Differential Analysis](#Module-3.-Differential-analysis-)</font>




## <B>Common tools of the trade</B>
***
### <B>- Explaination of tools</B>
|Tool|Website|Info|
|----|------|-----|
|BWA|http://bio-bwa.sourceforge.net/bwa.shtml| Genomic Sequence Read alignment tool |
|PICARD|https://broadinstitute.github.io/picard/| Toolkit for BAM file manipulation
|SAMTOOLS|http://www.htslib.org/doc/samtools.html| Toolkit for BAM file manipulation |
|BEDTOOLS|https://bedtools.readthedocs.io/en/latest/index.html | Toolkit for BED/BEDGRAPH file manipulation |
|MACS2|https://github.com/macs3-project/MACS | Enriched region identifier for ChIP-seq |
|UCSCtools|http://genome.ucsc.edu/goldenPath/help/bigWig.html| Manipulation of Wigs and Bed/Bedgraph to binary forms|

### <B>- Resource files</B>
|File name | File location| Source
|---|---|--|
|BWA INDEX|~/CourseData/EPI_data/Module1/BWA_index/| |
|Enhancer file.bed|~/CourseData/EPI_data/Module1/QC_resources/|download various state7 and merge https://egg2.wustl.edu/roadmap/data/byFileType/chromhmmSegmentations/ChmmModels/coreMarks/jointModel/final/|
|TSS.bed|~/CourseData/EPI_data/Module1/QC_resources/|Generated by downloading Ensemblv79 GTF convert to Bed +/-2kb of TSS. See https://www.biostars.org/p/56280/|
|HOX regions.bed|~/CourseData/EPI_data/Module1/QC_resources/|Generated by downloading Ensemblv79 GTF convert to Bed then selecting for HOX|
|Hg38 Black list regions|~/CourseData/EPI_data/Module1/QC_resources/|https://www.encodeproject.org/files/ENCFF356LFX/@@download/ENCFF356LFX.bed.gz|

### <B>- Resource files</B>
|File name | File location| Source
|---|---|--|
|CEMT Pooled Breast Basal|~/CourseData/EPI_data/Module1/CHEERC_resources|https://epigenomesportal.ca/tracks/CEEHRC/hg38/69069.CEEHRC.CEMT0035.H3K27ac.peak_calls.bigBed
|CEMT Pooled Breast Basal ||https://epigenomesportal.ca/tracks/CEEHRC/hg38/69067.CEEHRC.CEMT0035.H3K27ac.signal_unstranded.bigWig
|CEMT Pooled Breast Basal |~/CourseData/EPI_data/Module1/CHEERC_resources|https://epigenomesportal.ca/tracks/CEEHRC/hg38/69063.CEEHRC.CEMT0035.H3K27me3.peak_calls.bigBed
|CEMT Pooled Breast Basal ||https://epigenomesportal.ca/tracks/CEEHRC/hg38/69061.CEEHRC.CEMT0035.H3K27me3.signal_unstranded.bigWig
|CEMT Pooled Breast Basal |~/CourseData/EPI_data/Module1/CHEERC_resources|https://epigenomesportal.ca/tracks/CEEHRC/hg38/69054.CEEHRC.CEMT0035.H3K4me1.peak_calls.bigBed
|CEMT Pooled Breast Basal | |https://epigenomesportal.ca/tracks/CEEHRC/hg38/69052.CEEHRC.CEMT0035.H3K4me1.signal_unstranded.bigWig
|CEMT Pooled Breast Basal |~/CourseData/EPI_data/Module1/CHEERC_resources|https://epigenomesportal.ca/tracks/CEEHRC/hg38/69057.CEEHRC.CEMT0035.H3K4me3.peak_calls.bigBed
|CEMT Pooled Breast Basal | |https://epigenomesportal.ca/tracks/CEEHRC/hg38/69055.CEEHRC.CEMT0035.H3K4me3.signal_unstranded.bigWig
|CEMT Pooled Breast Basal | |https://epigenomesportal.ca/tracks/CEEHRC/hg38/69070.CEEHRC.CEMT0035.Input.signal_unstranded.bigWig
|CEMT Pooled Breast Stromal|~/CourseData/EPI_data/Module1/CHEERC_resources|https://epigenomesportal.ca/tracks/CEEHRC/hg38/69088.CEEHRC.CEMT0036.H3K27ac.peak_calls.bigBed
|CEMT Pooled Breast Stromal| |https://epigenomesportal.ca/tracks/CEEHRC/hg38/69086.CEEHRC.CEMT0036.H3K27ac.signal_unstranded.bigWig
|CEMT Pooled Breast Stromal|~/CourseData/EPI_data/Module1/CHEERC_resources|https://epigenomesportal.ca/tracks/CEEHRC/hg38/69082.CEEHRC.CEMT0036.H3K27me3.peak_calls.bigBed
|CEMT Pooled Breast Stromal| |https://epigenomesportal.ca/tracks/CEEHRC/hg38/69080.CEEHRC.CEMT0036.H3K27me3.signal_unstranded.bigWig
|CEMT Pooled Breast Stromal|~/CourseData/EPI_data/Module1/CHEERC_resources|https://epigenomesportal.ca/tracks/CEEHRC/hg38/69073.CEEHRC.CEMT0036.H3K4me1.peak_calls.bigBed
|CEMT Pooled Breast Stromal| |https://epigenomesportal.ca/tracks/CEEHRC/hg38/69071.CEEHRC.CEMT0036.H3K4me1.signal_unstranded.bigWig
|CEMT Pooled Breast Stromal|~/CourseData/EPI_data/Module1/CHEERC_resources|https://epigenomesportal.ca/tracks/CEEHRC/hg38/69076.CEEHRC.CEMT0036.H3K4me3.peak_calls.bigBed
|CEMT Pooled Breast Stromal| |https://epigenomesportal.ca/tracks/CEEHRC/hg38/69074.CEEHRC.CEMT0036.H3K4me3.signal_unstranded.bigWig
|CEMT Pooled Breast Stromal| |https://epigenomesportal.ca/tracks/CEEHRC/hg38/69089.CEEHRC.CEMT0036.Input.signal_unstranded.bigWig
|CEMT Pooled Breast Luminal|~/CourseData/EPI_data/Module1/CHEERC_resources|https://epigenomesportal.ca/tracks/CEEHRC/hg38/69107.CEEHRC.CEMT0037.H3K27ac.peak_calls.bigBed
|CEMT Pooled Breast Luminal| |https://epigenomesportal.ca/tracks/CEEHRC/hg38/69105.CEEHRC.CEMT0037.H3K27ac.signal_unstranded.bigWig
|CEMT Pooled Breast Luminal|~/CourseData/EPI_data/Module1/CHEERC_resources|https://epigenomesportal.ca/tracks/CEEHRC/hg38/69101.CEEHRC.CEMT0037.H3K27me3.peak_calls.bigBed
|CEMT Pooled Breast Luminal| |https://epigenomesportal.ca/tracks/CEEHRC/hg38/69099.CEEHRC.CEMT0037.H3K27me3.signal_unstranded.bigWig
|CEMT Pooled Breast Luminal|~/CourseData/EPI_data/Module1/CHEERC_resources|https://epigenomesportal.ca/tracks/CEEHRC/hg38/69092.CEEHRC.CEMT0037.H3K4me1.peak_calls.bigBed
|CEMT Pooled Breast Luminal| |https://epigenomesportal.ca/tracks/CEEHRC/hg38/69090.CEEHRC.CEMT0037.H3K4me1.signal_unstranded.bigWig
|CEMT Pooled Breast Luminal|~/CourseData/EPI_data/Module1/CHEERC_resources|https://epigenomesportal.ca/tracks/CEEHRC/hg38/69095.CEEHRC.CEMT0037.H3K4me3.peak_calls.bigBed
|CEMT Pooled Breast Luminal| |https://epigenomesportal.ca/tracks/CEEHRC/hg38/69093.CEEHRC.CEMT0037.H3K4me3.signal_unstranded.bigWig
|CEMT Pooled Breast Luminal| |https://epigenomesportal.ca/tracks/CEEHRC/hg38/69108.CEEHRC.CEMT0037.Input.signal_unstranded.bigWig
|CEMT Pooled Breast Luminal Progenitor|~/CourseData/EPI_data/Module1/CHEERC_resources|https://epigenomesportal.ca/tracks/CEEHRC/hg38/69126.CEEHRC.CEMT0038.H3K27ac.peak_calls.bigBed
|CEMT Pooled Breast Luminal Progenitor| |https://epigenomesportal.ca/tracks/CEEHRC/hg38/69124.CEEHRC.CEMT0038.H3K27ac.signal_unstranded.bigWig
|CEMT Pooled Breast Luminal Progenitor|~/CourseData/EPI_data/Module1/CHEERC_resources|https://epigenomesportal.ca/tracks/CEEHRC/hg38/69120.CEEHRC.CEMT0038.H3K27me3.peak_calls.bigBed
|CEMT Pooled Breast Luminal Progenitor| |https://epigenomesportal.ca/tracks/CEEHRC/hg38/69118.CEEHRC.CEMT0038.H3K27me3.signal_unstranded.bigWig
|CEMT Pooled Breast Luminal Progenitor|~/CourseData/EPI_data/Module1/CHEERC_resources|https://epigenomesportal.ca/tracks/CEEHRC/hg38/69111.CEEHRC.CEMT0038.H3K4me1.peak_calls.bigBed
|CEMT Pooled Breast Luminal Progenitor| |https://epigenomesportal.ca/tracks/CEEHRC/hg38/69109.CEEHRC.CEMT0038.H3K4me1.signal_unstranded.bigWig
|CEMT Pooled Breast Luminal Progenitor|~/CourseData/EPI_data/Module1/CHEERC_resources|https://epigenomesportal.ca/tracks/CEEHRC/hg38/69114.CEEHRC.CEMT0038.H3K4me3.peak_calls.bigBed
|CEMT Pooled Breast Luminal Progenitor| |https://epigenomesportal.ca/tracks/CEEHRC/hg38/69112.CEEHRC.CEMT0038.H3K4me3.signal_unstranded.bigWiig
|CEMT Pooled Breast Luminal Progenitor| |https://epigenomesportal.ca/tracks/CEEHRC/hg38/69127.CEEHRC.CEMT0038.Input.signal_unstranded.bigWig

## <B>Module 3. Differential analysis </B>
***

### <B>Setup</B>
```Shell
mkdir analysis
```

#### <B>- Comparing regions</B>

##### Code:
```Shell
MCF10A_H3K27ac=~/workspace/peaks/MCF10A_H3K27ac_peaks.blacklistRemoved.narrowPeak
Basal_H3K27ac=~/CourseData/EPI_data/Module1/CEEHRC_resources/Basal_69069.CEEHRC.CEMT0035.H3K27ac.peak.bed


mkdir -p ~/workspace/analysis
bedtools intersect -u -a ${MCF10A_H3K27ac} -b ${Basal_H3K27ac} | wc -l
bedtools intersect -v -a ${MCF10A_H3K27ac} -b ${Basal_H3K27ac} | cut -f1-3 > ~/workspace/analysis/MCF10A_H3K27ac_unique.bed
bedtools intersect -v -b ${MCF10A_H3K27ac} -a ${Basal_H3K27ac} | wc -l
wc -l ~/workspace/analysis/MCF10A_H3K27ac_unique.bed
```

##### Output:
```Shell
18355
107949
14029
```
##### Comments:
```Shell
bedtools intersect \ #Tool intersects regions from A onto B to perform a function
-u \ # A region of A could intersect the multiple regions of B, thus we do not want to over count and only report it once
-a ${MCF10A_H3K27ac} \ # Depending on the operation bedtools will only provide the output relative to "A", thus if we want "B" we would have to switch the command order
-b ${Basal_H3K27ac} \ # Comperator
| wc -l # Line count

bedtools intersect -u -a ${MCF10A_H3K27ac} -b ${Basal_H3K27ac} | wc -l # Reports number of unique regions in A that are in common with B
bedtools intersect -v -a ${MCF10A_H3K27ac} -b ${Basal_H3K27ac} | wc -l # Reports number of regions that are not found in B
bedtools intersect -v -b ${MCF10A_H3K27ac} -a ${Basal_H3K27ac} | wc -l # We swapped A with B to report regions unique to B
```

##### Peak Count Results:
|Mark| |MCF10A|Basal|Luminal|LuminalProgenitor|Stromal|
|--|--|--|--|--|--|--|
|H3K27ac||32384|126386|76203|102867|149943|
|||A-B|14029|13643|13267|14560|
|||AuB|18355|18741|19117|17824|
|||B-A|107949|59645|84630|131251|
|H3K27me3||184034|314557|383257|344099|361087|
|||A-B|87203|86349|93197|84175|
|||AuB|96831|97685|90837|99859|
|||B-A|140374|198506|178309|177294|
|H3K4me1||141821|186957|170000|177095|204897|
|||A-B|83813|88651|86570|89159|
|||AuB|58008|53170|55251|52662|
|||B-A|91024|84548|87121|114524|
|H3K4me3||27632|26230|23216|29104|24974|
|||A-B|9578|10599|9281|10532|
|||AuB|18054|17033|18351|17100|
|||B-A|7993|6370|10669|7789|



## <B>Cheat sheet </B>
***


#### <B>Dedup Bams</B>
##### Code:
<code style="background:black;color:black">
mkdir -p ~/workspace/peaks
for mark in H3K27ac H3K4me3 H3K27me3 H3K4me1 Input;
do
    bam=~/CourseData/EPI_data/Module1/MCF10A_resources/MCF10A_${mark}.bam
    dedup=~/workspace/peaks/MCF10A_${mark}.dedup.bam
    samtools view -@4 ${bam} -bh -q10 -F1028 -o ${dedup}
done 
</code>

#### <B>Peak Calling all the marks</B>
##### Code:
<code style="background:black;color:black">
input_frag=~/workspace/peaks/MCF10A_Input.dedup.bam
for narrow_mark in H3K27ac H3K4me3;
do
    treatment_frag=~/workspace/peaks/MCF10A_${narrow_mark}.dedup.bam
    macs3 callpeak -t ${treatment_frag} -c ${input_frag} -f BAMPE -g hs -n MCF10A_${narrow_mark} --keep-dup all --outdir ~/workspace/peaks/ --bdg 1> ~/workspace/peaks/${narrow_mark}.out.log 2> ~/workspace/peaks/${narrow_mark}.err.log 
done
for broad_mark in H3K4me1 H3K27me3;
do
    treatment_frag=~/workspace/peaks/MCF10A_${broad_mark}.dedup.bam
    macs3 callpeak -t ${treatment_frag} -c ${input_frag} -f BAMPE -g hs -n MCF10A_${broad_mark} --keep-dup all --outdir ~/workspace/peaks/ --bdg --broad 1> ~/workspace/peaks/${broad_mark}.out.log 2> ~/workspace/peaks/${broad_mark}.err.log 
done
</code>

#### <B>Blacklist Filtering all Peaks</B>
##### Code:
<code style="background:black;color:black">
blacklist=~/CourseData/EPI_data/Module1/QC_resources/hg38_blacklist.bed
for narrow_mark in H3K27ac H3K4me3;
do
    bedtools intersect -v -a ~/workspace/peaks/MCF10A_${narrow_mark}_peaks.narrowPeak -b ${blacklist} > ~/workspace/peaks/MCF10A_${narrow_mark}_peaks.blacklistRemoved.narrowPeak
    bedtools intersect -u -a ~/workspace/peaks/MCF10A_${narrow_mark}_peaks.narrowPeak -b ${blacklist} > ~/workspace/peaks/MCF10A_${narrow_mark}_peaks.blacklist.narrowPeak
done
for broad_mark in H3K4me1 H3K27me3;
do
    bedtools intersect -v -a ~/workspace/peaks/MCF10A_${broad_mark}_peaks.broadPeak -b ${blacklist} > ~/workspace/peaks/MCF10A_${broad_mark}_peaks.blacklistRemoved.broadPeak
    bedtools intersect -u -a ~/workspace/peaks/MCF10A_${broad_mark}_peaks.broadPeak -b ${blacklist} > ~/workspace/peaks/MCF10A_${broad_mark}_peaks.blacklist.broadPeak
done
</code>

#### <B>BigWig converting all your *treat_pileup.bdg</B>
##### Code:
<code style="background:black;color:black">
mkdir -p ~/workspace/{bigBed,bigWig}
chrom_sizes=~/CourseData/EPI_data/Module1/QC_resources/hg38.chrom.sizes
for mark in H3K27ac H3K4me3 H3K27me3 H3K4me1
do
    input_bedgraph=~/workspace/peaks/MCF10A_${mark}_treat_pileup.bdg
    output_bigwig=~/workspace/bigWig/MCF10A_${mark}_treat_pileup.bw
    sort -k1,1 -k2,2n ${input_bedgraph} > ~/workspace/bigWig/tmp
    bedGraphToBigWig ~/workspace/bigWig/tmp ${chrom_sizes} ${output_bigwig}
    rm ~/workspace/bigWig/tmp
done
</code>

<code style="background:black;color:black">
sample="MCF10A_H3K27ac"
input_bedgraph=~/workspace/peaks/${sample}_control_lambda.bdg
output_bigwig=~/workspace/bigWig/${sample}_control_lambda.bw
chrom_sizes=~/CourseData/EPI_data/Module1/QC_resources/hg38.chrom.sizes
sort -k1,1 -k2,2n ${input_bedgraph} > ~/workspace/bigWig/tmp
bedGraphToBigWig ~/workspace/bigWig/tmp ${chrom_sizes} ${output_bigwig}
rm ~/workspace/bigWig/tmp
</code>


#### <B>BigBed converting all your *Peaks</B>
##### Code:
<code style="background:black;color:black">
mkdir -p ~/workspace/{bigBed,bigWig}
chrom_sizes=~/CourseData/EPI_data/Module1/QC_resources/hg38.chrom.sizes
for narrow_mark in H3K27ac H3K4me3;
do
    input_bed=~/workspace/peaks/MCF10A_${narrow_mark}_peaks.blacklistRemoved.narrowPeak
    output_bigbed=~/workspace/bigBed/MCF10A_${narrow_mark}_peaks.blacklistRemoved.bb
    chrom_sizes=~/CourseData/EPI_data/Module1/QC_resources/hg38.chrom.sizes
    cut -f1-3 ${input_bed} | sort -k1,1 -k2,2n  > ~/workspace/bigBed/tmp
    bedToBigBed ~/workspace/bigBed/tmp ${chrom_sizes} ${output_bigbed}
done
for broad_mark in H3K4me1 H3K27me3;
do
    input_bed=~/workspace/peaks/MCF10A_${broad_mark}_peaks.blacklistRemoved.broadPeak
    output_bigbed=~/workspace/bigBed/MCF10A_${broad_mark}_peaks.blacklistRemoved.bb
    chrom_sizes=~/CourseData/EPI_data/Module1/QC_resources/hg38.chrom.sizes
    cut -f1-3 ${input_bed} | sort -k1,1 -k2,2n  > ~/workspace/bigBed/tmp
    bedToBigBed ~/workspace/bigBed/tmp ${chrom_sizes} ${output_bigbed}
done
</code>

#### <B>Total read counts per BAM</B>
##### Code:
<code style="background:black;color:black">
for narrow_mark in H3K27ac H3K4me3;
do
    bam=~/CourseData/EPI_data/Module1/MCF10A_resources/MCF10A_${narrow_mark}.bam
    peak=~/workspace/peaks/MCF10A_${narrow_mark}_peaks.blacklistRemoved.narrowPeak
    echo total reads in ${narrow_mark} $(samtools view -@4 ${bam} -q10 -F1028 -c )
done
for broad_mark in H3K4me1 H3K27me3;
do
    bam=~/CourseData/EPI_data/Module1/MCF10A_resources/MCF10A_${broad_mark}.bam
    peak=~/workspace/peaks/MCF10A_${broad_mark}_peaks.blacklistRemoved.narrowPeak
    echo total reads in ${broad_mark} $(samtools view -@4 ${bam} -bh -q10 -F1028 -c )
done
</code>

#### <B>Reads in Enriched region per BAM</B>
##### Code:
<code style="background:black;color:black">
for narrow_mark in H3K27ac H3K4me3;
do
    bam=~/CourseData/EPI_data/Module1/MCF10A_resources/MCF10A_${narrow_mark}.bam
    peak=~/workspace/peaks/MCF10A_${narrow_mark}_peaks.blacklistRemoved.narrowPeak
    echo enriched reads in ${narrow_mark} $(samtools view -@4 ${bam} -q10 -F1028 -c -L ${peak})
done
for broad_mark in H3K4me1 H3K27me3;
do
    bam=~/CourseData/EPI_data/Module1/MCF10A_resources/MCF10A_${broad_mark}.bam
    peak=~/workspace/peaks/MCF10A_${broad_mark}_peaks.blacklistRemoved.broadPeak
    echo enriched reads in ${broad_mark} $(samtools view -@4 ${bam} -bh -q10 -F1028 -c -L ${peak})
done
</code>

#### <B>Reads in Domain regions per BAM</B>
##### Code:
<code style="background:black;color:black">
for file in $(ls ~/CourseData/EPI_data/Module1/QC_resources/* | grep bed | grep -v blacklist);
do
    for bam in $(ls ~/CourseData/EPI_data/Module1/MCF10A_resources/*bam );
    do
        echo $(basename ${bam}) $(basename ${file}) $(samtools view -@4 -q 10 -F 1028 ${bam} -L $file -c )
    done
done
</code>

#### <B>Counts total peaks per file</B>
##### Code:
<code style="background:black;color:black">
for file in $(ls ~/workspace/peaks/*.blacklistRemoved.*Peak);
do
    wc -l $file
done
for file in $(ls ~/CourseData/EPI_data/Module1/CEEHRC_resources/*.bed);
do
    wc -l $file
done
</code>

#### <B>Counts total peaks per overlap</B>
##### Code:
<code style="background:black;color:black">
for histone_mark in H3K27ac H3K27me3 H3K4me3 H3K4me1;
do
    for fileA in $(ls ~/workspace/peaks/*${histone_mark}*.blacklistRemoved*Peak);
    do
        for fileB in $(ls ~/CourseData/EPI_data/Module1/CEEHRC_resources/*${histone_mark}*.bed);
        do
            echo $(basename $fileA) $(basename $fileB) AuB $(bedtools intersect -u -a $fileA -b $fileB | wc -l)
            echo $(basename $fileA) $(basename $fileB) A-B $(bedtools intersect -v -a $fileA -b $fileB | wc -l)
            echo $(basename $fileA) $(basename $fileB) B-A $(bedtools intersect -v -b $fileA -a $fileB | wc -l)
        done
    done
done
</code>
