---
layout: tutorial_page
permalink: /epigenomics_2018_prework
title: Epigenomic Data Analysis Prework
header1: Workshop Pages for Students
header2: Epigenomic Data Analysis 2018
image: /site_images/CBW_Epigenome-data_icon.jpg
home: https://bioinformaticsdotca.github.io	/epigenomics_2018
---

## Pre-readings, Pre-work, and Laptop Setup   

### Install These Tools Before the Workshop

**A robust text editor.**

For Windows/PC - notepad++ (http://notepad-plus-plus.org/).
For Linux - gedit (http://projects.gnome.org/gedit/).
For Mac – TextWrangler (http://www.barebones.com/products/textwrangler/download.html)

**A file decompression tool.**
For Windows/PC – 7zip (http://www.7-zip.org/).
For Linux – already there.
For Mac – already there.

**A robust internet browser** such as Chrome (Internet Explorer is not recommended). Firefox and Safari are also fine.

**A PDF viewer** (Adobe Acrobat or equivalent).

**Java** -The visualization program that we will be using (IGV) requires Java. Check if you have Java installed at https://www.java.com/verify/ and download it if it is not.

**Integrative Genomics Viewer 2.3 (IGV)**
With Java installed, go to http://www.broadinstitute.org/igv/ and register in order to get access to the download page. Once there, click on the appropriate launch button that matches the amount of memory available on your laptop (if you have space, 1.2GB is good, more is better).
Chrome: Does not launch "java webstart" files by default. Instead, the launch buttons below will download a "jnlp" file. This should appear in the lower left corner of the browser. Double-click the downloaded file to run.
Windows users: To run with more than 1.2 GB you must install 64-bit Java. In general trying to launch with more memory than your OS/Java combination supports will result in the obscure error "could not create virtual machine".

**SSH client** - Mac and Linux users already have the command line ssh program that can be run from the terminal. On a MAC, the Terminal window is located in Applications/Utilities. Tip: add the terminal window to your dock so it is easy to open when needed.  For Windows users, please download PuTTY: http://www.chiark.greenend.org.uk/~sgtatham/putty/download.html

**SCP/SFTP client** - We will be moving data from the servers to the student laptops for visualization. Mac and Linux users already have the command line applications scp and sftp. For Windows users, please install WinSCP: http://winscp.net/eng/download.php

### Pre-workshop Readings

* [Introduction to High-Performance Computing](http://insidehpc.com/hpc-basic-training/what-is-hpc/)

* [Introduction to Compute Canada](https://www.computecanada.ca/research-portal/accessing-resources/)

* [Introduction to epigenomics](http://www.ncbi.nlm.nih.gov/pmc/articles/PMC3080743/)

* [Additional reading on ChIP-Seq](http://www.ncbi.nlm.nih.gov/pmc/articles/PMC3053263/)

* [Additional reading on Methylation analysis](http://www.ncbi.nlm.nih.gov/pubmed/22986265/) 

### Do These Tutorials Before the Workshop:

* [UNIX Bootcamp](http://rik.smith-unna.com/command_line_bootcamp/?id=9xnbkx6eaof)  

* [Unix cheat sheet](http://www.rain.org/~mkummel/unix.html)

* [IGV Tutorial](https://bioinformaticsdotca.github.io/BiCG_2017_module2_igv): Review how to use IGV Genome Browser if you have not used this tool before.

* [UCSC Genome Browser tutorial](http://www.ncbi.nlm.nih.gov/pubmed/18514479)

* [FastQC Introduction and examples](http://www.bioinformatics.babraham.ac.uk/projects/fastqc/Help/)  

Please note that these instructions may change prior to the workshop.  
