<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Meet Your Faculty | Epigenomic Analysis 2023</title>
  <meta name="description" content="Workshop website for the {{YEAR}} {{WORKSHOP_NAME}} Canadian Bioinformatics Workshop" />
  <meta name="generator" content="bookdown 0.43 and GitBook 2.6.7" />

  <meta property="og:title" content="Meet Your Faculty | Epigenomic Analysis 2023" />
  <meta property="og:type" content="book" />
  <meta property="og:image" content="https://bioinformaticsdotca.github.io/EPI_2023/img/bioinformatics_logo.png" />
  <meta property="og:description" content="Workshop website for the {{YEAR}} {{WORKSHOP_NAME}} Canadian Bioinformatics Workshop" />
  <meta name="github-repo" content="bioinformaticsdotca/EPI_2023" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Meet Your Faculty | Epigenomic Analysis 2023" />
  
  <meta name="twitter:description" content="Workshop website for the {{YEAR}} {{WORKSHOP_NAME}} Canadian Bioinformatics Workshop" />
  <meta name="twitter:image" content="https://bioinformaticsdotca.github.io/EPI_2023/img/bioinformatics_logo.png" />

<meta name="author" content="Faculty: Guillaume Bourque, Martin Hirst, David Bujold, Jose Hector Galvez, Edmund Su, Mareike Janiak, Michelle Brazas, Nia Hughes, and Zhibin Lu" />



  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  <link rel="shortcut icon" href="img/favicon.ico" type="image/x-icon" />
<link rel="prev" href="index.html"/>
<link rel="next" href="data-and-compute-setup.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.5.1/css/all.min.css">
<script type="application/ld+json">
{
  "@context": "https://schema.org",
  "@type": "LearningResource",
  "name": "{{WORKSHOP_NAME}}",
  "description": "The complete set of tutorials, slides, and practical exercises for the {{WORKSHOP_NAME}} {{YEAR}} Canadian Bioinformatics Workshop, hosted on GitHub Pages.",
  "url": "https://bioinformaticsdotca.github.io/{{WORKSHOP_CODE}}",
  "learningResourceType": ["tutorial", "slides", "exercise"],
  "inLanguage": "en",
  "license": "https://creativecommons.org/licenses/by/4.0/",
  "provider": {
    "@type": "Organization",
    "name": "Bioinformatics.ca",
    "url": "https://bioinformatics.ca"
  },
  "keywords": "",
  "isPartOf": {
    "@type": "Course",
    "@id": "{{URL}}"
  }
}
</script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
  
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<a href="https://bioinformaticsdotca.github.io/">
 <img src="img/bioinformatics_logo.png" width="90%" alt="bioinformatics.ca logo" style="display:block; margin-left:auto; margin-right:auto; margin-top:15px;">
 </a>
<li><center><strong><a href="./"><br>Epigenomic Analysis<br><br></a></strong></center></li>
<a href="./">
 <img src="img/CBW_Epigenome-data_icon.png" width="70%" alt="Epigenomic Analysis logo" style="display:block; margin-left:auto; margin-right:auto; ">
 </a>
<br>

<li class="divider"></li>
<li class="part"><span><b>I Introduction</b></span></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Workshop Info</a>
<ul>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#pre-work"><i class="fa fa-check"></i>Pre-work</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#class-photo"><i class="fa fa-check"></i>Class Photo</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#schedule"><i class="fa fa-check"></i>Schedule</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="meet-your-faculty.html"><a href="meet-your-faculty.html"><i class="fa fa-check"></i>Meet Your Faculty</a></li>
<li class="chapter" data-level="" data-path="data-and-compute-setup.html"><a href="data-and-compute-setup.html"><i class="fa fa-check"></i>Data and Compute Setup</a></li>
<li class="part"><span><b>II Modules</b></span></li>
<li class="chapter" data-level="" data-path="module-1-introduction-to-chip-sequencing-and-analysis.html"><a href="module-1-introduction-to-chip-sequencing-and-analysis.html"><i class="fa fa-check"></i>Module 1: Introduction to ChIP Sequencing and Analysis</a>
<ul>
<li class="chapter" data-level="" data-path="module-1-introduction-to-chip-sequencing-and-analysis.html"><a href="module-1-introduction-to-chip-sequencing-and-analysis.html#lecture"><i class="fa fa-check"></i>Lecture</a></li>
<li class="chapter" data-level="" data-path="module-1-introduction-to-chip-sequencing-and-analysis.html"><a href="module-1-introduction-to-chip-sequencing-and-analysis.html#lab"><i class="fa fa-check"></i>Lab</a>
<ul>
<li class="chapter" data-level="" data-path="module-1-introduction-to-chip-sequencing-and-analysis.html"><a href="module-1-introduction-to-chip-sequencing-and-analysis.html#breakdown-of-markdown-file"><i class="fa fa-check"></i>Breakdown of Markdown File</a></li>
<li class="chapter" data-level="" data-path="module-1-introduction-to-chip-sequencing-and-analysis.html"><a href="module-1-introduction-to-chip-sequencing-and-analysis.html#module-1.-alignment"><i class="fa fa-check"></i>Module 1. Alignment</a></li>
<li class="chapter" data-level="" data-path="module-1-introduction-to-chip-sequencing-and-analysis.html"><a href="module-1-introduction-to-chip-sequencing-and-analysis.html#server-resources"><i class="fa fa-check"></i>Server resources</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="" data-path="module-2-alignment-peak-calling-and-visualization.html"><a href="module-2-alignment-peak-calling-and-visualization.html"><i class="fa fa-check"></i>Module 2: Alignment, Peak Calling, and Visualization</a>
<ul>
<li class="chapter" data-level="" data-path="module-2-alignment-peak-calling-and-visualization.html"><a href="module-2-alignment-peak-calling-and-visualization.html#lecture-1"><i class="fa fa-check"></i>Lecture</a></li>
<li class="chapter" data-level="" data-path="module-2-alignment-peak-calling-and-visualization.html"><a href="module-2-alignment-peak-calling-and-visualization.html#lab-1"><i class="fa fa-check"></i>Lab</a>
<ul>
<li class="chapter" data-level="" data-path="module-2-alignment-peak-calling-and-visualization.html"><a href="module-2-alignment-peak-calling-and-visualization.html#breakdown-of-markdown-file-1"><i class="fa fa-check"></i>Breakdown of Markdown File</a></li>
<li class="chapter" data-level="" data-path="module-2-alignment-peak-calling-and-visualization.html"><a href="module-2-alignment-peak-calling-and-visualization.html#module-2.-peak-calling"><i class="fa fa-check"></i>Module 2. Peak Calling</a></li>
<li class="chapter" data-level="" data-path="module-2-alignment-peak-calling-and-visualization.html"><a href="module-2-alignment-peak-calling-and-visualization.html#server-resources-1"><i class="fa fa-check"></i>Server resources</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="" data-path="module-3-chip-seq-differential-analysis.html"><a href="module-3-chip-seq-differential-analysis.html"><i class="fa fa-check"></i>Module 3: ChIP-seq Differential Analysis</a>
<ul>
<li class="chapter" data-level="" data-path="module-3-chip-seq-differential-analysis.html"><a href="module-3-chip-seq-differential-analysis.html#lecture-2"><i class="fa fa-check"></i>Lecture</a></li>
<li class="chapter" data-level="" data-path="module-3-chip-seq-differential-analysis.html"><a href="module-3-chip-seq-differential-analysis.html#lab-2"><i class="fa fa-check"></i>Lab</a>
<ul>
<li class="chapter" data-level="" data-path="module-3-chip-seq-differential-analysis.html"><a href="module-3-chip-seq-differential-analysis.html#breakdown-of-markdown-file-2"><i class="fa fa-check"></i>Breakdown of markdown file</a></li>
<li class="chapter" data-level="" data-path="module-3-chip-seq-differential-analysis.html"><a href="module-3-chip-seq-differential-analysis.html#module-3---differential-analysis"><i class="fa fa-check"></i>Module 3 - Differential Analysis</a></li>
<li class="chapter" data-level="" data-path="module-3-chip-seq-differential-analysis.html"><a href="module-3-chip-seq-differential-analysis.html#server-resources-2"><i class="fa fa-check"></i>Server resources</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="" data-path="module-4-whole-genome-bisulfite-sequencing-and-analysis.html"><a href="module-4-whole-genome-bisulfite-sequencing-and-analysis.html"><i class="fa fa-check"></i>Module 4: Whole-Genome Bisulfite Sequencing and Analysis</a>
<ul>
<li class="chapter" data-level="" data-path="module-4-whole-genome-bisulfite-sequencing-and-analysis.html"><a href="module-4-whole-genome-bisulfite-sequencing-and-analysis.html#lecture-3"><i class="fa fa-check"></i>Lecture</a></li>
<li class="chapter" data-level="" data-path="module-4-whole-genome-bisulfite-sequencing-and-analysis.html"><a href="module-4-whole-genome-bisulfite-sequencing-and-analysis.html#lab-3"><i class="fa fa-check"></i>Lab</a>
<ul>
<li class="chapter" data-level="" data-path="module-4-whole-genome-bisulfite-sequencing-and-analysis.html"><a href="module-4-whole-genome-bisulfite-sequencing-and-analysis.html#introduction"><i class="fa fa-check"></i>1. Introduction</a></li>
<li class="chapter" data-level="" data-path="module-4-whole-genome-bisulfite-sequencing-and-analysis.html"><a href="module-4-whole-genome-bisulfite-sequencing-and-analysis.html#mapping-tutorial"><i class="fa fa-check"></i>2. Mapping Tutorial</a></li>
<li class="chapter" data-level="" data-path="module-4-whole-genome-bisulfite-sequencing-and-analysis.html"><a href="module-4-whole-genome-bisulfite-sequencing-and-analysis.html#differential-methylation-analysis-in-methylkit"><i class="fa fa-check"></i>3. Differential Methylation Analysis in MethylKit</a></li>
<li class="chapter" data-level="" data-path="module-4-whole-genome-bisulfite-sequencing-and-analysis.html"><a href="module-4-whole-genome-bisulfite-sequencing-and-analysis.html#differential-methylation-analysis-in-dss"><i class="fa fa-check"></i>4. Differential Methylation Analysis in DSS</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="" data-path="module-5-downstream-analysis-and-online-tools.html"><a href="module-5-downstream-analysis-and-online-tools.html"><i class="fa fa-check"></i>Module 5: Downstream Analysis and Online Tools</a>
<ul>
<li class="chapter" data-level="" data-path="module-5-downstream-analysis-and-online-tools.html"><a href="module-5-downstream-analysis-and-online-tools.html#lecture-4"><i class="fa fa-check"></i>Lecture</a></li>
<li class="chapter" data-level="" data-path="module-5-downstream-analysis-and-online-tools.html"><a href="module-5-downstream-analysis-and-online-tools.html#lab-4"><i class="fa fa-check"></i>Lab</a>
<ul>
<li class="chapter" data-level="" data-path="module-5-downstream-analysis-and-online-tools.html"><a href="module-5-downstream-analysis-and-online-tools.html#introduction-1"><i class="fa fa-check"></i>Introduction</a></li>
<li class="chapter" data-level="" data-path="module-5-downstream-analysis-and-online-tools.html"><a href="module-5-downstream-analysis-and-online-tools.html#ihec-data-portal"><i class="fa fa-check"></i>IHEC Data Portal</a></li>
<li class="chapter" data-level="" data-path="module-5-downstream-analysis-and-online-tools.html"><a href="module-5-downstream-analysis-and-online-tools.html#predicting-motifs-with-homer"><i class="fa fa-check"></i>Predicting Motifs with HOMER</a></li>
<li class="chapter" data-level="" data-path="module-5-downstream-analysis-and-online-tools.html"><a href="module-5-downstream-analysis-and-online-tools.html#looking-for-go-terms-enrichment-with-great"><i class="fa fa-check"></i>Looking for GO Terms Enrichment with GREAT</a></li>
<li class="chapter" data-level="" data-path="module-5-downstream-analysis-and-online-tools.html"><a href="module-5-downstream-analysis-and-online-tools.html#going-back-to-homer-results"><i class="fa fa-check"></i>Going back to HOMER Results</a></li>
</ul></li>
</ul></li>
<li class="divider"></li>
<h1 id="sponsors">Sponsors</h1>
<center>
<<a href="https://genomecanada.ca/"><img src="img/sponsors/genomecanada.png" height="70" style="display:inline-block;"  style="padding-bottom: 20px" alt="Genome Canada logo.">
<a href="https://www.ontariogenomics.ca/"><img src="img/sponsors/ontariogenomics.png" height="70" style="display:inline-block;" style="padding-bottom: 20px;" alt="Ontario Genomics logo.">
<br><br>
<a href="https://oicr.on.ca/"><img src="img/sponsors/oicr.png" width=25% style="display:inline-block;" style="padding-bottom: 20px" alt="Ontario Institute for Cancer Research logo.">
<a href="https://www.hpcforhealth.ca/"><img src="img/sponsors/hpc4health.png" width=60% style="display:inline-block;" style="padding-bottom: 20px "alt="HPC4Health logo.">
<br><br>
<a href="https://computationalgenomics.ca/"><img src="img/sponsors/c3g_Web.png" style="display:inline-block;" width=35% style="padding-bottom: 20px" alt="Canadian Centre for Computational Genomcis logo.">

<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li></center>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Epigenomic Analysis 2023</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="meet-your-faculty" class="section level1 hasAnchor">
<h1>Meet Your Faculty<a href="meet-your-faculty.html#meet-your-faculty" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<div id="guillaume-bourque" class="section level4 hasAnchor">
<h4>Guillaume Bourque<a href="meet-your-faculty.html#guillaume-bourque" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p><img src="img/faculty/guillaume-bourque.png" width="200" alt="Photo of Guillaume Bourque"></p>
<blockquote>
<p>Professor, McGill University <br>
Director of Bioinformatics, Genome Quebec Innovation Centre <br>
Director, Canadian Centre of Computational Genomics <br>
Director, McGill initiative for Computational Medicine <br></p>
</blockquote>
<p>Dr. Bourque research interests are in comparative and functional
genomics with a special emphasis on applications of next-generation sequencing technologies.
His lab develops advanced tools and scalable computational infrastructure to enable large-scale
applied research projects.</p>
</div>
<div id="martin-hirst" class="section level4 hasAnchor">
<h4>Martin Hirst<a href="meet-your-faculty.html#martin-hirst" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p><img src="img/faculty/martin-hirst.png" width="200" alt="Photo of Martin Hirst"></p>
<blockquote>
<p>Distinguished Scientist, BC Cancer <br>
Professor, Department of Microbiology &amp; Immunology <br>
Director, Michael Smith Laboratories <br>
Michael Smith Laboratories</p>
</blockquote>
<p>Dr. Hirst’s research focuses on understanding epigenetic dysfunction in
cancer and his laboratory develops experimental and computational tools to characterize normal
and transformed cell types down to the single cell level. He applies these tools to explore the
epigenomic states of normal and transformed cell types to discover and exploit therapeutic
vulnerabilities.</p>
</div>
<div id="david-bujold" class="section level4 hasAnchor">
<h4>David Bujold<a href="meet-your-faculty.html#david-bujold" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p><img src="img/faculty/david-bujold.png" width="200" alt="Photo of David Bujold"></p>
<blockquote>
<p>Bioinformatics Manager, Data Unit <br>
Canadian Centre of Computational Genomics</p>
</blockquote>
<p>He joined the McGill Epigenomic Data Coordination Center at McGill in
2012 to tackle challenges related to epigenomics, and has since developed
many data management and discovery solutions, including the IHEC Data Portal. Other projects
of interest include CanDIG and EpiShare, platforms to make genomic and epigenomic data
under controlled access more accessible, while maintaining study participants’ privacy.</p>
</div>
<div id="jose-hector-galvez" class="section level4 hasAnchor">
<h4>Jose Hector Galvez<a href="meet-your-faculty.html#jose-hector-galvez" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p><img src="img/faculty/jose-hector-galvez.png" width="200" alt="Photo of Jose Hector Galvez"></p>
<blockquote>
<p>Bioinformatics Manager, Tech Dev Unit <br>
Canadian Centre of Computational Genomics</p>
</blockquote>
<p>As a Bioinformatics Specialist in the Research and Development team,
Jose Hector is involved in maintaining, documenting, and upgrading
the RNA-seq pipelines in GenPipes. He also collaborates in several research projects, mostly
focusing on transcriptomics, genome assembly, and epigenomics.</p>
</div>
<div id="edmund-su" class="section level4 hasAnchor">
<h4>Edmund Su<a href="meet-your-faculty.html#edmund-su" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p><img src="img/faculty/edmund-su.png" width="200" alt="Photo of Edmund Su"></p>
<blockquote>
<p>Bioinformatician <br>
Ontario Institute for Cancer Research</p>
</blockquote>
<p>Edmund is a bioinformatician within the genome informatics team at OICR,
where he provides technical knowledge and expertise in genomic analysis.
His main focus is developing pipelines and data wrangling for ICGC-ARGO
(International Cancer Genome Consortium - Accelerating Research in Genomic Oncology).</p>
</div>
<div id="mareike-janiak" class="section level4 hasAnchor">
<h4>Mareike Janiak<a href="meet-your-faculty.html#mareike-janiak" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p><img src="img/faculty/mareike-janiak.png" width="200" alt="Photo of Mareike Janiak"></p>
<blockquote>
<p>Bioinformatics Analyst, Tech Dev Unit <br>
Canadian Centre of Computational Genomics</p>
</blockquote>
<p>As a Bioinformatics Analyst in the TechDev team, Mareike is responsible
for maintaining pipelines in GenPipes, testing and developing new pipelines
for the community, and responding to user requests. Prior to joining C3G,
she worked as both a field and computational biologist, with her doctoral and postdoctoral
research focusing on mammalian comparative genomics, phylogenomics, and metagenomics,
most often in primates.</p>
</div>
<div id="michelle-brazas" class="section level4 hasAnchor">
<h4>Michelle Brazas<a href="meet-your-faculty.html#michelle-brazas" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p><img src="img/faculty/michelle-brazas.jpg" width="200" alt="Photo of Michelle Brazas"></p>
<blockquote>
<p>Acting Scientific Director <br>
Canadian Bioinformatics Workshops (CBW) <br>
Toronto, ON, CA</p>
<p>— <a href="mailto:support@bioinformatics.ca" class="email">support@bioinformatics.ca</a></p>
</blockquote>
<p>Dr. Michelle Brazas is the Associate Director for Adaptive Oncology at the
Ontario Institute for Cancer Research (OICR), and acting Scientific Director at Bioinformatics.ca.
Previously, Dr. Brazas was the Program Manager for Bioinformatics.ca and a faculty member in
Biotechnology at BCIT. Michelle co-founded and runs the Toronto Bioinformatics User Group
(TorBUG) now in its 11th season, and plays an active role in the International Society of
Computational Biology where she sits on the Board of Directors and Executive Board.</p>
</div>
<div id="nia-hughes" class="section level4 hasAnchor">
<h4>Nia Hughes<a href="meet-your-faculty.html#nia-hughes" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p><img src="img/faculty/nia-hughes.jpeg" width="200" alt="Photo of Nia Hughes"></p>
<blockquote>
<p>Program Manager, Bioinformatics.ca <br>
Ontario Institute for Cancer Research <br>
Toronto, ON, Canada</p>
<p>— <a href="mailto:nia.hughes@oicr.on.ca" class="email">nia.hughes@oicr.on.ca</a></p>
</blockquote>
<p>Nia is the Program Manager for Bioinformatics.ca, where she coordinates the Canadian
Bioinformatics Workshop Series. Prior to starting at OICR, she completed her M.Sc. in
Bioinformatics from the University of Guelph in 2020 before working there as a bioinformatician
studying epigenetic and transcriptomic patterns across maize varieties.</p>
</div>
<div id="zhibin-lu" class="section level4 hasAnchor">
<h4>Zhibin Lu<a href="meet-your-faculty.html#zhibin-lu" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p><img src="img/faculty/zhibin-lu.png" width="200" alt="Photo of Zhibin Lu"></p>
<blockquote>
<p>HPC and Bioinformatics Services Manager at Princess Margaret <br>
Cancer Centre, University Health Network <br>
Bioinformatics and HPC Core, UHN <br>
MaRS Centre, PMCRT 11-707 <br>
101 College St <br>
Toronto ON M5G 1L7</p>
<p>— <a href="mailto:zhibin@gmail.com" class="email">zhibin@gmail.com</a> <br>
<a href="https://bhpc.uhnresearch.ca/">https://bhpc.uhnresearch.ca/</a></p>
</blockquote>
<p>Zhibin Lu is a senior manager at University Health Network Digital. He is responsible for UHN
HPC operations and scientific software. He manages two HPC clusters at UHN, including
system administration, user management, and maintenance of bioinformatics tools for
HPC4health. He is also skilled in Next-Gen sequence data analysis and has developed and</p>
<p>maintained bioinformatics pipelines at the Bioinformatics and HPC Core. He is a member of the
Digital Research Alliance of Canada Bioinformatics National Team and Scheduling National
Team.</p>

</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="index.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="data-and-compute-setup.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
  "sharing": {
    "github": false,
    "facebook": false,
    "twitter": false,
    "linkedin": false,
    "weibo": false,
    "instapaper": false,
    "vk": false,
    "whatsapp": false,
    "all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
  },
  "fontsettings": {
    "theme": "white",
    "family": "sans",
    "size": 2
  },
  "edit": {
    "link": "https://github.com/bioinformaticsdotca/EPI_2023/blob/main/001-faculty.Rmd",
    "text": "Edit"
  },
  "history": {
    "link": null,
    "text": null
  },
  "view": {
    "link": null,
    "text": null
  },
  "download": ["_main.pdf", "_main.epub"],
  "search": {
    "engine": "fuse",
    "options": null
  },
  "toc": {
    "collapse": "subsection"
  }
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.9/latest.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
