<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Module 3: ChIP-seq Differential Analysis | Epigenomic Analysis 2023</title>
  <meta name="description" content="Workshop website for the {{YEAR}} {{WORKSHOP_NAME}} Canadian Bioinformatics Workshop" />
  <meta name="generator" content="bookdown 0.43 and GitBook 2.6.7" />

  <meta property="og:title" content="Module 3: ChIP-seq Differential Analysis | Epigenomic Analysis 2023" />
  <meta property="og:type" content="book" />
  <meta property="og:image" content="https://bioinformaticsdotca.github.io/EPI_2023/img/bioinformatics_logo.png" />
  <meta property="og:description" content="Workshop website for the {{YEAR}} {{WORKSHOP_NAME}} Canadian Bioinformatics Workshop" />
  <meta name="github-repo" content="bioinformaticsdotca/EPI_2023" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Module 3: ChIP-seq Differential Analysis | Epigenomic Analysis 2023" />
  
  <meta name="twitter:description" content="Workshop website for the {{YEAR}} {{WORKSHOP_NAME}} Canadian Bioinformatics Workshop" />
  <meta name="twitter:image" content="https://bioinformaticsdotca.github.io/EPI_2023/img/bioinformatics_logo.png" />

<meta name="author" content="Faculty: Guillaume Bourque, Martin Hirst, David Bujold, Jose Hector Galvez, Edmund Su, Mareike Janiak, Michelle Brazas, Nia Hughes, and Zhibin Lu" />



  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  <link rel="shortcut icon" href="img/favicon.ico" type="image/x-icon" />
<link rel="prev" href="module-2-alignment-peak-calling-and-visualization.html"/>
<link rel="next" href="module-4-whole-genome-bisulfite-sequencing-and-analysis.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.5.1/css/all.min.css">
<script type="application/ld+json">
{
  "@context": "https://schema.org",
  "@type": "LearningResource",
  "name": "{{WORKSHOP_NAME}}",
  "description": "The complete set of tutorials, slides, and practical exercises for the {{WORKSHOP_NAME}} {{YEAR}} Canadian Bioinformatics Workshop, hosted on GitHub Pages.",
  "url": "https://bioinformaticsdotca.github.io/{{WORKSHOP_CODE}}",
  "learningResourceType": ["tutorial", "slides", "exercise"],
  "inLanguage": "en",
  "license": "https://creativecommons.org/licenses/by/4.0/",
  "provider": {
    "@type": "Organization",
    "name": "Bioinformatics.ca",
    "url": "https://bioinformatics.ca"
  },
  "keywords": "",
  "isPartOf": {
    "@type": "Course",
    "@id": "{{URL}}"
  }
}
</script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
  
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<a href="https://bioinformaticsdotca.github.io/">
 <img src="img/bioinformatics_logo.png" width="90%" alt="bioinformatics.ca logo" style="display:block; margin-left:auto; margin-right:auto; margin-top:15px;">
 </a>
<li><center><strong><a href="./"><br>Epigenomic Analysis<br><br></a></strong></center></li>
<a href="./">
 <img src="img/CBW_Epigenome-data_icon.png" width="70%" alt="Epigenomic Analysis logo" style="display:block; margin-left:auto; margin-right:auto; ">
 </a>
<br>

<li class="divider"></li>
<li class="part"><span><b>I Introduction</b></span></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Workshop Info</a>
<ul>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#pre-work"><i class="fa fa-check"></i>Pre-work</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#class-photo"><i class="fa fa-check"></i>Class Photo</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#schedule"><i class="fa fa-check"></i>Schedule</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="meet-your-faculty.html"><a href="meet-your-faculty.html"><i class="fa fa-check"></i>Meet Your Faculty</a></li>
<li class="chapter" data-level="" data-path="data-and-compute-setup.html"><a href="data-and-compute-setup.html"><i class="fa fa-check"></i>Data and Compute Setup</a></li>
<li class="part"><span><b>II Modules</b></span></li>
<li class="chapter" data-level="" data-path="module-1-introduction-to-chip-sequencing-and-analysis.html"><a href="module-1-introduction-to-chip-sequencing-and-analysis.html"><i class="fa fa-check"></i>Module 1: Introduction to ChIP Sequencing and Analysis</a>
<ul>
<li class="chapter" data-level="" data-path="module-1-introduction-to-chip-sequencing-and-analysis.html"><a href="module-1-introduction-to-chip-sequencing-and-analysis.html#lecture"><i class="fa fa-check"></i>Lecture</a></li>
<li class="chapter" data-level="" data-path="module-1-introduction-to-chip-sequencing-and-analysis.html"><a href="module-1-introduction-to-chip-sequencing-and-analysis.html#lab"><i class="fa fa-check"></i>Lab</a>
<ul>
<li class="chapter" data-level="" data-path="module-1-introduction-to-chip-sequencing-and-analysis.html"><a href="module-1-introduction-to-chip-sequencing-and-analysis.html#breakdown-of-markdown-file"><i class="fa fa-check"></i>Breakdown of Markdown File</a></li>
<li class="chapter" data-level="" data-path="module-1-introduction-to-chip-sequencing-and-analysis.html"><a href="module-1-introduction-to-chip-sequencing-and-analysis.html#module-1.-alignment"><i class="fa fa-check"></i>Module 1. Alignment</a></li>
<li class="chapter" data-level="" data-path="module-1-introduction-to-chip-sequencing-and-analysis.html"><a href="module-1-introduction-to-chip-sequencing-and-analysis.html#server-resources"><i class="fa fa-check"></i>Server resources</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="" data-path="module-2-alignment-peak-calling-and-visualization.html"><a href="module-2-alignment-peak-calling-and-visualization.html"><i class="fa fa-check"></i>Module 2: Alignment, Peak Calling, and Visualization</a>
<ul>
<li class="chapter" data-level="" data-path="module-2-alignment-peak-calling-and-visualization.html"><a href="module-2-alignment-peak-calling-and-visualization.html#lecture-1"><i class="fa fa-check"></i>Lecture</a></li>
<li class="chapter" data-level="" data-path="module-2-alignment-peak-calling-and-visualization.html"><a href="module-2-alignment-peak-calling-and-visualization.html#lab-1"><i class="fa fa-check"></i>Lab</a>
<ul>
<li class="chapter" data-level="" data-path="module-2-alignment-peak-calling-and-visualization.html"><a href="module-2-alignment-peak-calling-and-visualization.html#breakdown-of-markdown-file-1"><i class="fa fa-check"></i>Breakdown of Markdown File</a></li>
<li class="chapter" data-level="" data-path="module-2-alignment-peak-calling-and-visualization.html"><a href="module-2-alignment-peak-calling-and-visualization.html#module-2.-peak-calling"><i class="fa fa-check"></i>Module 2. Peak Calling</a></li>
<li class="chapter" data-level="" data-path="module-2-alignment-peak-calling-and-visualization.html"><a href="module-2-alignment-peak-calling-and-visualization.html#server-resources-1"><i class="fa fa-check"></i>Server resources</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="" data-path="module-3-chip-seq-differential-analysis.html"><a href="module-3-chip-seq-differential-analysis.html"><i class="fa fa-check"></i>Module 3: ChIP-seq Differential Analysis</a>
<ul>
<li class="chapter" data-level="" data-path="module-3-chip-seq-differential-analysis.html"><a href="module-3-chip-seq-differential-analysis.html#lecture-2"><i class="fa fa-check"></i>Lecture</a></li>
<li class="chapter" data-level="" data-path="module-3-chip-seq-differential-analysis.html"><a href="module-3-chip-seq-differential-analysis.html#lab-2"><i class="fa fa-check"></i>Lab</a>
<ul>
<li class="chapter" data-level="" data-path="module-3-chip-seq-differential-analysis.html"><a href="module-3-chip-seq-differential-analysis.html#breakdown-of-markdown-file-2"><i class="fa fa-check"></i>Breakdown of markdown file</a></li>
<li class="chapter" data-level="" data-path="module-3-chip-seq-differential-analysis.html"><a href="module-3-chip-seq-differential-analysis.html#module-3---differential-analysis"><i class="fa fa-check"></i>Module 3 - Differential Analysis</a></li>
<li class="chapter" data-level="" data-path="module-3-chip-seq-differential-analysis.html"><a href="module-3-chip-seq-differential-analysis.html#server-resources-2"><i class="fa fa-check"></i>Server resources</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="" data-path="module-4-whole-genome-bisulfite-sequencing-and-analysis.html"><a href="module-4-whole-genome-bisulfite-sequencing-and-analysis.html"><i class="fa fa-check"></i>Module 4: Whole-Genome Bisulfite Sequencing and Analysis</a>
<ul>
<li class="chapter" data-level="" data-path="module-4-whole-genome-bisulfite-sequencing-and-analysis.html"><a href="module-4-whole-genome-bisulfite-sequencing-and-analysis.html#lecture-3"><i class="fa fa-check"></i>Lecture</a></li>
<li class="chapter" data-level="" data-path="module-4-whole-genome-bisulfite-sequencing-and-analysis.html"><a href="module-4-whole-genome-bisulfite-sequencing-and-analysis.html#lab-3"><i class="fa fa-check"></i>Lab</a>
<ul>
<li class="chapter" data-level="" data-path="module-4-whole-genome-bisulfite-sequencing-and-analysis.html"><a href="module-4-whole-genome-bisulfite-sequencing-and-analysis.html#introduction"><i class="fa fa-check"></i>1. Introduction</a></li>
<li class="chapter" data-level="" data-path="module-4-whole-genome-bisulfite-sequencing-and-analysis.html"><a href="module-4-whole-genome-bisulfite-sequencing-and-analysis.html#mapping-tutorial"><i class="fa fa-check"></i>2. Mapping Tutorial</a></li>
<li class="chapter" data-level="" data-path="module-4-whole-genome-bisulfite-sequencing-and-analysis.html"><a href="module-4-whole-genome-bisulfite-sequencing-and-analysis.html#differential-methylation-analysis-in-methylkit"><i class="fa fa-check"></i>3. Differential Methylation Analysis in MethylKit</a></li>
<li class="chapter" data-level="" data-path="module-4-whole-genome-bisulfite-sequencing-and-analysis.html"><a href="module-4-whole-genome-bisulfite-sequencing-and-analysis.html#differential-methylation-analysis-in-dss"><i class="fa fa-check"></i>4. Differential Methylation Analysis in DSS</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="" data-path="module-5-downstream-analysis-and-online-tools.html"><a href="module-5-downstream-analysis-and-online-tools.html"><i class="fa fa-check"></i>Module 5: Downstream Analysis and Online Tools</a>
<ul>
<li class="chapter" data-level="" data-path="module-5-downstream-analysis-and-online-tools.html"><a href="module-5-downstream-analysis-and-online-tools.html#lecture-4"><i class="fa fa-check"></i>Lecture</a></li>
<li class="chapter" data-level="" data-path="module-5-downstream-analysis-and-online-tools.html"><a href="module-5-downstream-analysis-and-online-tools.html#lab-4"><i class="fa fa-check"></i>Lab</a>
<ul>
<li class="chapter" data-level="" data-path="module-5-downstream-analysis-and-online-tools.html"><a href="module-5-downstream-analysis-and-online-tools.html#introduction-1"><i class="fa fa-check"></i>Introduction</a></li>
<li class="chapter" data-level="" data-path="module-5-downstream-analysis-and-online-tools.html"><a href="module-5-downstream-analysis-and-online-tools.html#ihec-data-portal"><i class="fa fa-check"></i>IHEC Data Portal</a></li>
<li class="chapter" data-level="" data-path="module-5-downstream-analysis-and-online-tools.html"><a href="module-5-downstream-analysis-and-online-tools.html#predicting-motifs-with-homer"><i class="fa fa-check"></i>Predicting Motifs with HOMER</a></li>
<li class="chapter" data-level="" data-path="module-5-downstream-analysis-and-online-tools.html"><a href="module-5-downstream-analysis-and-online-tools.html#looking-for-go-terms-enrichment-with-great"><i class="fa fa-check"></i>Looking for GO Terms Enrichment with GREAT</a></li>
<li class="chapter" data-level="" data-path="module-5-downstream-analysis-and-online-tools.html"><a href="module-5-downstream-analysis-and-online-tools.html#going-back-to-homer-results"><i class="fa fa-check"></i>Going back to HOMER Results</a></li>
</ul></li>
</ul></li>
<li class="divider"></li>
<h1 id="sponsors">Sponsors</h1>
<center>
<<a href="https://genomecanada.ca/"><img src="img/sponsors/genomecanada.png" height="70" style="display:inline-block;"  style="padding-bottom: 20px" alt="Genome Canada logo.">
<a href="https://www.ontariogenomics.ca/"><img src="img/sponsors/ontariogenomics.png" height="70" style="display:inline-block;" style="padding-bottom: 20px;" alt="Ontario Genomics logo.">
<br><br>
<a href="https://oicr.on.ca/"><img src="img/sponsors/oicr.png" width=25% style="display:inline-block;" style="padding-bottom: 20px" alt="Ontario Institute for Cancer Research logo.">
<a href="https://www.hpcforhealth.ca/"><img src="img/sponsors/hpc4health.png" width=60% style="display:inline-block;" style="padding-bottom: 20px "alt="HPC4Health logo.">
<br><br>
<a href="https://computationalgenomics.ca/"><img src="img/sponsors/c3g_Web.png" style="display:inline-block;" width=35% style="padding-bottom: 20px" alt="Canadian Centre for Computational Genomcis logo.">

<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li></center>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Epigenomic Analysis 2023</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="module-3-chip-seq-differential-analysis" class="section level1 hasAnchor">
<h1>Module 3: ChIP-seq Differential Analysis<a href="module-3-chip-seq-differential-analysis.html#module-3-chip-seq-differential-analysis" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<div id="lecture-2" class="section level2 hasAnchor">
<h2>Lecture<a href="module-3-chip-seq-differential-analysis.html#lecture-2" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<iframe width="640" height="360" src="https://www.youtube.com/embed/Ja3TbSw2Lao?si=pyK5vOwJsdNiauhP" title="YouTube video player" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture; web-share" referrerpolicy="strict-origin-when-cross-origin" allowfullscreen>
</iframe>
<p><br></p>
<iframe src="https://drive.google.com/file/d/1LPc4t8mdc1RypIWCaWqJKhjju_4xkGQ0/preview" width="640" height="480" allow="autoplay">
</iframe>
</div>
<div id="lab-2" class="section level2 hasAnchor">
<h2>Lab<a href="module-3-chip-seq-differential-analysis.html#lab-2" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<div id="breakdown-of-markdown-file-2" class="section level3 hasAnchor">
<h3>Breakdown of markdown file<a href="module-3-chip-seq-differential-analysis.html#breakdown-of-markdown-file-2" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<ul>
<li>At the start of each step, the intention will declare.</li>
<li>this is then follow by a code block</li>
</ul>
<p><strong>Code:</strong></p>
<pre><code>Like this! This is the main code to run for the step.

Additionally the code block will include a header to indicate what environment to the run code for example:
###Shell###
pwd

###R###
getwd()</code></pre>
<ul>
<li>explaining for commands will be broken down following code block</li>
<li><code>pwd</code> &amp; <code>getwd()</code>
<ul>
<li>see your work directory</li>
</ul></li>
<li>sprinkled throughout will also include comments</li>
</ul>
<div class="callout-blue">
<span class="callout-icon"><i class="fas fa-circle-info" aria-hidden="true"></i></span>
<div class="callout-content">
<div class="callout-title-static">
<p>Note</p>
</div>
<p>Important points and considerations will also be raised as so.</p>
</div>
</div>
</div>
<div id="module-3---differential-analysis" class="section level3 hasAnchor">
<h3>Module 3 - Differential Analysis<a href="module-3-chip-seq-differential-analysis.html#module-3---differential-analysis" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<div id="step1a-copy-bigwig-resources" class="section level4 hasAnchor">
<h4>Step1A: Copy bigWig resources<a href="module-3-chip-seq-differential-analysis.html#step1a-copy-bigwig-resources" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<ul>
<li>we’ll load some bigWigs to compare with</li>
</ul>
<p><strong>Code</strong></p>
<pre><code>###Shell###
cp ~/CourseData/EPI_data/module123/encode_bigWig/*H3K4me3* ~/workspace/module123/bigWig/
cp ~/CourseData/EPI_data/module123/encode_bigBed/*H3K4me3* ~/workspace/module123/bigBed/
cp ~/CourseData/EPI_data/module123/triplicates/bigWig/* ~/workspace/module123/bigWig/</code></pre>
</div>
<div id="step1b-using-bedtools-to-compare-marks" class="section level4 hasAnchor">
<h4>Step1B: Using Bedtools to compare marks<a href="module-3-chip-seq-differential-analysis.html#step1b-using-bedtools-to-compare-marks" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<ul>
<li>We’ll explore how to use bedtools to compare MCF10A histone marks and the interpretation of results</li>
</ul>
<p><strong>Code</strong></p>
<pre><code>###Shell###
MCF10A_H3K27ac=~/workspace/module123/peaks/MCF10A_H3K27ac_peaks.blacklistRemoved.narrowPeak
MCF10A_H3K27me3=~/workspace/module123/peaks/MCF10A_H3K27me3_peaks.blacklistRemoved.broadPeak
MCF10A_H3K4me3=~/workspace/module123/peaks/MCF10A_H3K4me3_peaks.blacklistRemoved.narrowPeak

bedtools intersect -u -a ${MCF10A_H3K27ac} -b ${MCF10A_H3K27me3} | wc -l
bedtools intersect -u -a ${MCF10A_H3K27ac} -b ${MCF10A_H3K4me3} | wc -l</code></pre>
<ul>
<li><code>bedtools intersect -u -a ${MCF10A_H3K27ac} -b ${MCF10A_H3K27me3} | wc -l</code>
<ul>
<li>results in an intersect of <code>6/988</code></li>
</ul></li>
<li><code>bedtools intersect -u -a ${MCF10A_H3K27ac} -b ${MCF10A_H3K4me3} | wc -l</code>
<ul>
<li>results in an intersect of <code>789/988</code></li>
</ul></li>
<li>what do we know about the relationship of H3K27ac vs H3K27me3 vs H3K4me3?
<ul>
<li>H3K27ac and H3K27me3 tend to be antagonistic, hence the very small intersect</li>
<li>H3K27ac co-occurs with H3K4me3 at promoters, hence the larger intersect</li>
</ul></li>
</ul>
</div>
<div id="step1c-using-bedtools-to-compare-samples" class="section level4 hasAnchor">
<h4>Step1C: Using Bedtools to compare samples<a href="module-3-chip-seq-differential-analysis.html#step1c-using-bedtools-to-compare-samples" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<ul>
<li>We’ll demonstrate how to do comparisons on mass against ENCODE breast data and interpret results</li>
</ul>
<p><strong>Code</strong></p>
<pre><code>###Shell###
basal_H3K27ac=~/CourseData/EPI_data/module123/encode_bed/basal.H3K27ac.peak_calls.bed
luminal_H3K27ac=~/CourseData/EPI_data/module123/encode_bed/luminal.H3K27ac.peak_calls.bed
stromal_H3K27ac=~/CourseData/EPI_data/module123/encode_bed/stromal.H3K27ac.peak_calls.bed
lp_H3K27ac=~/CourseData/EPI_data/module123/encode_bed/lp.H3K27ac.peak_calls.bed
MCF10A_H3K27ac=~/workspace/module123/peaks/MCF10A_H3K27ac_peaks.blacklistRemoved.narrowPeak

paste \
&lt;(ls ~/CourseData/EPI_data/module123/encode_bed/*H3K4me3* | xargs -I {} sh -c &quot;bedtools intersect -u -a ~/workspace/module123/peaks/MCF10A_H3K4me3_peaks.blacklistRemoved.narrowPeak -b {} | wc -l&quot;) \
&lt;(ls ~/CourseData/EPI_data/module123/encode_bed/*H3K27ac* |  xargs -I {} sh -c &quot;bedtools intersect -u -a ~/workspace/module123/peaks/MCF10A_H3K27ac_peaks.blacklistRemoved.narrowPeak -b {} | wc -l&quot;) \
&lt;(ls ~/CourseData/EPI_data/module123/encode_bed/*H3K27me3* |  xargs -I {} sh -c &quot;bedtools intersect -u -a ~/workspace/module123/peaks/MCF10A_H3K27me3_peaks.blacklistRemoved.broadPeak -b {} | wc -l&quot;)</code></pre>
<ul>
<li><code>paste \ &lt;(ls CourseData/module123/bed/*H3K4me3* | xargs -I {} sh -c "bedtools intersect -u -a workspace/module123/peaks/MCF10A_H3K4me3_peaks.blacklistRemoved.narrowPeak -b {} | wc -l") \ &lt;(ls CourseData/module123/bed/*H3K27ac* |  xargs -I {} sh -c "bedtools intersect -u -a workspace/module123/peaks/MCF10A_H3K27ac_peaks.blacklistRemoved.narrowPeak -b {} | wc -l")</code>
<ul>
<li><code>paste</code> lets us aggregate our results where each column is the output from the command within <code>&lt;(COMMAND)</code></li>
<li>Each <code>&lt;(COMMAND)</code>, contains the following <code>LOOKUP | FORLOOP intersect and count</code></li>
<li><code>\</code> let’s continue our command on another line</li>
</ul></li>
<li>Intersect numbers:</li>
</ul>
<table>
<thead>
<tr class="header">
<th></th>
<th>H3K4me3</th>
<th>H3K27ac</th>
<th>H3K27me3</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>MCF10A</td>
<td>1406</td>
<td>988</td>
<td>4797</td>
</tr>
<tr class="even">
<td>Intersecting Basal</td>
<td>1039</td>
<td>656</td>
<td>2420</td>
</tr>
<tr class="odd">
<td>Intersecting Luminal Progenitor</td>
<td>1099</td>
<td>778</td>
<td>2496</td>
</tr>
<tr class="even">
<td>Intersecting Luminal</td>
<td>1024</td>
<td>766</td>
<td>2430</td>
</tr>
<tr class="odd">
<td>Intersecting Stromal</td>
<td>978</td>
<td>717</td>
<td>2604</td>
</tr>
</tbody>
</table>
<ul>
<li>MCF10A is luminal progenitor like, how is that relationship reflect in the epigenetic landscape?
<ul>
<li>higher amount of intersect in permissive marks of H3K4me3 and H3K27ac</li>
</ul></li>
</ul>
</div>
<div id="step1d-using-bedtools-and-pipe" class="section level4 hasAnchor">
<h4>Step1D: Using Bedtools and pipe<a href="module-3-chip-seq-differential-analysis.html#step1d-using-bedtools-and-pipe" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<ul>
<li>We’ll demonstrate advanced queries by piping and doing multiple bedtool queries</li>
</ul>
<p><strong>Code:</strong></p>
<pre><code>###Shell###
basal_H3K4me3=~/CourseData/EPI_data/module123/encode_bed/basal.H3K4me3.peak_calls.bed
luminal_H3K4me3=~/CourseData/EPI_data/module123/encode_bed/luminal.H3K4me3.peak_calls.bed
stromal_H3K4me3=~/CourseData/EPI_data/module123/encode_bed/stromal.H3K4me3.peak_calls.bed
lp_H3K4me3=~/CourseData/EPI_data/module123/encode_bed/lp.H3K4me3.peak_calls.bed
MCF10A_H3K4me3=~/workspace/module123/peaks/MCF10A_H3K4me3_peaks.blacklistRemoved.narrowPeak

bedtools intersect -u -a ${MCF10A_H3K4me3} -b ${lp_H3K4me3} | bedtools intersect -v -a stdin -b ${basal_H3K4me3} ${luminal_H3K4me3} ${stromal_H3K4me3} | wc -l

bedtools intersect -u -a ${MCF10A_H3K4me3} -b ${lp_H3K4me3} |\
bedtools intersect -u -a stdin -b ${basal_H3K4me3} |\
bedtools intersect -u -a stdin -b ${luminal_H3K4me3} |\
bedtools intersect -u -a stdin -b ${stromal_H3K4me3} | wc -l</code></pre>
<ul>
<li><code>bedtools intersect -u -a ${MCF10A_H3K4me3} -b ${lp_H3K4me3} | bedtools intersect -v -a stdin -b ${basal_H3K4me3} ${luminal_H3K4me3} ${stromal_H3K4me3}| wc -l</code>
<ul>
<li>The command can be broken down into the following : <code>Common with LP | Not found in Basal OR Luminal OR stromal</code></li>
<li><code>stdin</code> takes the results from half of our command and utilizies as input in the next</li>
<li>Other tools may have a similar function of <code>stdin</code>. Check documentation first.</li>
</ul></li>
</ul>
<p><img src="img/MCF10A-pLP-mBLS.png" /></p>
<ul>
<li><code>bedtools intersect -u -a ${MCF10A_H3K4me3} -b ${lp_H3K4me3} |  bedtools intersect -u -a stdin -b  {basal_H3K4me3} | bedtools intersect -u -a stdin -b ${luminal_H3K4me3} | bedtools intersect -u -a stdin -b ${stromal_H3K4me3}  | wc -l</code>
<ul>
<li>The command can be broken down into the following : <code>MCF10A Common with LP | Common with Basal | Common with Luminal | Stromal</code></li>
</ul></li>
</ul>
<p><img src="img/MCF10A-pLPMBLS.png" /></p>
</div>
<div id="step1e-using-bedtools-to-compare-binary-conditionsmodels" class="section level4 hasAnchor">
<h4>Step1E: Using Bedtools to compare binary conditions/models<a href="module-3-chip-seq-differential-analysis.html#step1e-using-bedtools-to-compare-binary-conditionsmodels" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<ul>
<li>We’ll explore how to use bedtools to compare binary conditions and possible interpretations</li>
</ul>
<p><strong>Code:</strong></p>
<pre><code>###Shell###
condA_rep1=~/CourseData/EPI_data/module123/triplicates/peaks/CondA.Rep1_peaks.narrowPeak  
condB_rep1=~/CourseData/EPI_data/module123/triplicates/peaks/CondB.Rep1_peaks.narrowPeak
condA_rep2=~/CourseData/EPI_data/module123/triplicates/peaks/CondA.Rep2_peaks.narrowPeak  
condB_rep2=~/CourseData/EPI_data/module123/triplicates/peaks/CondB.Rep2_peaks.narrowPeak
condA_rep3=~/CourseData/EPI_data/module123/triplicates/peaks/CondA.Rep3_peaks.narrowPeak
condB_rep3=~/CourseData/EPI_data/module123/triplicates/peaks/CondB.Rep3_peaks.narrowPeak

bedtools intersect -u -a ${condA_rep1} -b ${condA_rep2} | wc -l
bedtools intersect -u -a ${condA_rep1} -b ${condB_rep2} | wc -l 
bedtools intersect -v -a ${condA_rep1} -b ${condA_rep2} | wc -l
bedtools intersect -v -a ${condA_rep1} -b ${condB_rep2} | wc -l

bedtools intersect -u -a ${condA_rep1} -b ${condA_rep2} ${condA_rep3} | wc -l

bedtools intersect -u -a ${condA_rep1} -b ${condA_rep2} ${condA_rep3} -f 0.5 -F 0.5 | wc -l

bedtools intersect -wao -a ${condA_rep1} -b ${condA_rep2} ${condA_rep3} | head</code></pre>
<ul>
<li><code>bedtools intersect -u -a ${condA_rep1} -b ${condA_rep2} | wc -l</code>
<ul>
<li>counting the number of condA_rep1 peaks that intersect condA_rep2</li>
<li>returns <code>1191</code></li>
</ul></li>
<li><code>bedtools intersect -u -a ${condA_rep1} -b ${condB_rep2} | wc -l</code>
<ul>
<li>counting the number of condA_rep1 peaks that intersect condB_rep2</li>
<li>returns <code>1093</code></li>
</ul></li>
<li><code>bedtools intersect -v -a ${condA_rep1} -b ${condA_rep2} | wc -l</code>
<ul>
<li>counting the number of condA_rep1 peaks that do not intersect condA_rep2</li>
<li>return <code>50</code></li>
</ul></li>
<li><code>bedtools intersect -v -a ${condA_rep1} -b ${condB_rep2} | wc -l</code>
<ul>
<li>counting the number of condA_rep1 peaks that do not intersect condB_rep2</li>
<li>returns <code>148</code></li>
</ul></li>
<li>as expected our replicates of matching conditions have more in common</li>
<li><code>bedtools intersect -u -a ${condA_rep1} -b ${condA_rep2} ${condA_rep3} | wc -l</code>
<ul>
<li>counting the number of condA_rep1 peaks that intersect condA_rep2 or condA_rep3</li>
</ul></li>
<li><code>bedtools intersect -wao -a ${condA_rep1} -b ${condA_rep2} ${condA_rep3} | head</code>
<ul>
<li>specify <code>-wao</code> returns the original line of <code>${condA_rep1}</code> and the element it intersects</li>
<li>additionally adds an identify column for the database and number of base pairs overlapping</li>
</ul></li>
<li><code>bedtools intersect -u -a ${condA_rep1} -b ${condA_rep2} ${condA_rep3} -f 0.5 -F 0.5 | wc -l</code>
<ul>
<li>the flag <code>-f 0.5</code> adds the conditions that inte##rsects are only counted when 50% overlap of A occurs</li>
<li>the flag <code>-F 0.5</code> adds the conditions that intersects are only counted when 50% overlap of B occurs</li>
<li>if we remove one of the flags, how does the number change?</li>
<li>what if we wanted integer threshold instead of percentage?</li>
</ul></li>
<li><code>bedtools intersect -wao -a ${condA_rep1} -b ${condA_rep2} ${condA_rep3} | head</code>
<ul>
<li>specify <code>-wao</code> returns the original line of <code>${condA_rep1}</code> and the element it intersects</li>
<li>additionally adds an identify column for the database and number of base pairs overlapping</li>
</ul></li>
</ul>
</div>
<div id="step1f-other-useful-bedtool-functions" class="section level4 hasAnchor">
<h4>Step1F: Other useful bedtool functions<a href="module-3-chip-seq-differential-analysis.html#step1f-other-useful-bedtool-functions" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<ul>
<li>We’ll highlight other useful bedtool applications.</li>
</ul>
<p><strong>Code:</strong></p>
<pre><code>###Shell###
MCF10A_H3K27ac=~/workspace/module123/peaks/MCF10A_H3K27ac_peaks.blacklistRemoved.narrowPeak
TSS=~/workspace/module123/resources/hg38v79_genes_tss_2000.bed

bedtools closest -a ${MCF10A_H3K27ac} -b ${TSS} -d | head

###
condA_peaks=~/CourseData/EPI_data/module123/triplicates/peaks/CondA.Rep1_peaks.narrowPeak
condB_peaks=~/CourseData/EPI_data/module123/triplicates/peaks/CondB.Rep1_peaks.narrowPeak

cat ${condA_peaks} ${condB_peaks} | sort -k1,1 -k2,2n | bedtools merge -i stdin &gt; ~/workspace/module123/deeptools/merged_peaks.bed

####
methylation=~/workspace/module123/resources/example_methylation.bed

echo chr19 1 58617616 |\
sed &#39;s/ /\t/g&#39; |\
bedtools makewindows -w 50 -b stdin |\
awk &#39;BEGIN{{srand(1)}}{print $0&quot;\t&quot;rand()}&#39; \
&gt; ${methylation}

bedtools map -a ${MCF10A_H3K27ac} -b ${methylation} -c 4 -o median,count | head</code></pre>
<ul>
<li><code>sed 's/ /\t/g'</code> replace a <code></code> with <code>\t</code></li>
<li><code>bedtools closest -a ${MCF10A_H3K27ac} -b ${TSS} -d | head</code>
<ul>
<li>Identifies features in fileB that are closes to fileA</li>
<li>useful for mapping enhancers to their closest transcription start site</li>
<li><code>-d</code> will make the tool report the distance</li>
<li>in our example, if the distance is zero the H3K27ac instead reflects an activated promoter.</li>
</ul></li>
<li><code>cat ${condA_rep1} ${condA_rep2} ${condA_rep3} | sort -k1,1 -k2,2n | bedtools merge -i stdin | wc -l</code>
<ul>
<li>Pseudo code : <code>read peak files | sort peak files | merge peak files | line count</code></li>
<li><code>bedtools merge</code> takes elements specified in the input and merges the features if they intersect</li>
<li>behaviour can be modified to require a certain mount overlap or bookended features</li>
<li>compare how many peaks <code>cat ${condA_rep1} ${condA_rep2} ${condA_rep3}</code> starts off with</li>
<li>following merge how many peaks are left?</li>
</ul></li>
<li><code>echo chr19\\t1\\t58617616 | bedtools makewindows -w 50 -b stdin | awk 'BEGIN{srand(1);}{print $0"\t"rand()}'  &gt; ${metylation}</code>
<ul>
<li>Pseudo code : <code>simulate a bed file of chr19 | turn bedFile into 50bp bins | add random float value</code></li>
<li><code>echo chr19\\t1\\t58617616</code>
<ul>
<li>’<code>\\t</code> an escape character is uses to generate a tab</li>
</ul></li>
<li><code>bedtools makewindows -w 50 -b stdin</code>
<ul>
<li><code>-w 50</code> specifies our window size</li>
<li>can alternatively use <code>-n</code> to specified how many windows we want to divide our bedFile into</li>
</ul></li>
<li><code>awk 'BEGIN{srand(1);}{print $0"\t"rand()}'</code>
<ul>
<li>read our input (the 50bp window bed file of chr19) and generate a random float</li>
<li><code>srand(1)</code> set our seed number. Changing this will affect our pseudo random numbers</li>
<li><code>print $0"\t"rand()</code> as <code>awk</code> process line by line, print the current line (the windowed genomic coordiantes) and a random float number</li>
</ul></li>
<li><code>bedtools map -a ${MCF10A_H3K27ac} -b ${methylation} -c 4 -o median,count</code>
<ul>
<li><code>bedtools map</code> apply a function summarizing the values of fileB that intersect fileA</li>
<li>in our example for we’re looking at the methylation of H3K27ac peaks</li>
<li><code>-c 4</code> indicates which columns from fileB we’d like to use</li>
<li><code>-o median,count</code> return the median of col 4 and coutn the number of elements from fileB that intersected the particular element from fileA</li>
</ul></li>
</ul></li>
</ul>
</div>
<div id="step2-differential-peaks-utilizing-triplicates-and-diffbind" class="section level4 hasAnchor">
<h4>Step2: Differential peaks utilizing triplicates and DiffBind<a href="module-3-chip-seq-differential-analysis.html#step2-differential-peaks-utilizing-triplicates-and-diffbind" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<ul>
<li>We’ll perform analysis on mock MFC10A H3K4me3 data to get significant differential peaks for each condition. To do so, we’ll utilizing the <code>diffBind</code> package in R</li>
</ul>
<p><strong>Code :</strong></p>
<pre><code>###R###
library(DiffBind)
setwd(&quot;/home/ubuntu&quot;)
samples &lt;- read.csv(&quot;CourseData/EPI_data/module123/triplicates/triplicates.csv&quot;)
MCF10A &lt;- dba(sampleSheet=samples)
MCF10A &lt;- dba.count(MCF10A, bUseSummarizeOverlaps=TRUE)
dba.plotPCA(MCF10A, attributes=DBA_CONDITION,label=DBA_ID)
plot(MCF10A)
MCF10A &lt;- dba.contrast(MCF10A, categories=DBA_CONDITION)

MCF10A &lt;- dba.analyze(MCF10A, method=DBA_EDGER)

analyzed_peaks &lt;- dba.report(MCF10A, method=DBA_EDGER, fold=1)

dba.plotMA(MCF10A, bXY=TRUE , method=DBA_EDGER, fold=1)

write.table(analyzed_peaks, file=&quot;workspace/module123/diffBind/differential_peaks.tsv&quot;, sep=&quot;\t&quot;, quote=F, row.names=F, col.names=F)</code></pre>
<ul>
<li><code>library(DiffBind)</code>
<ul>
<li>we load R package <a href="https://bioconductor.org/packages/release/bioc/html/DiffBind.html">DiffBind</a></li>
</ul></li>
<li><code>setwd("/home/ubuntu")</code>
<ul>
<li>set our working directory</li>
</ul></li>
<li><code>read.csv("CourseData/EPI_data/module123/triplicates/triplicates.csv")</code>
<ul>
<li>read in our csv</li>
<li>let’s inspect the columns</li>
</ul></li>
<li><code>MCF10A &lt;- dba(sampleSheet=samples)</code>
<ul>
<li>read our samplesheet into the <code>dba</code> object that will be saved as <code>MCF10A</code></li>
</ul></li>
<li><code>MCF10A &lt;- dba.count(MCF10A, bUseSummarizeOverlaps=TRUE)</code>
<ul>
<li>count the number of fragments that intersect with peaks</li>
<li><code>bUseSummarizeOverlaps=TRUE</code> indicates the counting module to be used. <code>SummarizeOverlaps</code> comes from <a href="https://www.rdocumentation.org/packages/GenomicAlignments/versions/1.8.4/topics/summarizeOverlaps-methods">GenomicAlignments</a>.</li>
</ul></li>
<li><code>dba.plotPCA(MCF10A, attributes=DBA_CONDITION,label=DBA_ID)</code>
<ul>
<li>generate a principle component analysis using data from our object <code>MCF10A</code> where the annotations are <code>DBA_CONDITION</code> and the labelling is <code>DBA_ID</code></li>
</ul></li>
<li><code>plot(MCF10A)</code>
<ul>
<li>generate a heatmap with correlation and dendrogram</li>
<li>should note the correlation is based on score of overlap and not pearson and spearman, should recalculate</li>
</ul></li>
<li><code>MCF10A &lt;- dba.contrast(MCF10A, categories=DBA_CONDITION)</code>
<ul>
<li>declares what are the conditions for our differential groups</li>
<li><code>categories=DBA_CONDITION</code> the category we want to compare</li>
</ul></li>
<li><code>MCF10A &lt;- dba.analyze(MCF10A, method=DBA_EDGER)</code>
<ul>
<li>perform an analysis based on the <code>contrast</code> we previously established.</li>
<li><code>method=DBA_EDGER</code>, analysis engine is a library called <code>edgeR</code></li>
<li>note for our specific example <code>deseq2</code> does not work. <code>Deseq2</code> has a built in check for variablity which our synthetic dataset is lacking</li>
</ul></li>
<li><code>analyzed_peaks &lt;- dba.report(MCF10A, method=DBA_EDGER, fold=1)</code>
<ul>
<li>report the peaks identified by <code>DBA_EDGER</code> to be significant and have an absolute fold change <code>&gt;1</code></li>
</ul></li>
<li><code>dba.plotMA(MCF10A, bXY=TRUE , method=DBA_EDGER, fold=1)</code>
<ul>
<li>generates Scatter plot</li>
<li><code>method=DBA_EDGER</code> fetch results based on our previous analysis using <code>edgeR</code></li>
<li><code>bXY=TRUE</code> produces a scatter plot, <code>FALSE</code> produces a MA plot</li>
<li><code>fold=1</code> report differential positions that meet fold change threshold</li>
</ul></li>
<li><code>write.table(analyzed_peaks, file="workspace/module123/diffBind/differential_peaks.tsv", sep="\t", quote=F, row.names=F, col.names=T)</code>
<ul>
<li>save our differential peaks to a TSV</li>
<li><code>sep="\t"</code> the seperator to be used</li>
<li><code>col.names=T</code> include column names</li>
<li><code>row.names=F</code> include row names</li>
<li><code>quote=F</code> if we want to include quotations around values</li>
</ul></li>
</ul>
<p><img src="img/diffBind.png" /></p>
</div>
<div id="step3a-differential-peaks-utilizing-fold-change-and-significance---merged-peaks" class="section level4 hasAnchor">
<h4>Step3A: Differential peaks utilizing Fold change and significance - Merged peaks<a href="module-3-chip-seq-differential-analysis.html#step3a-differential-peaks-utilizing-fold-change-and-significance---merged-peaks" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<ul>
<li>Previously we performed differential analysis on triplicates, now let’s explore how to do so on two samples.</li>
<li>We’ll combine our two peak sets into an unified set.</li>
</ul>
<p><strong>Code:</strong></p>
<pre><code>###Shell###
condA_peaks=~/CourseData/EPI_data/module123/triplicates/peaks/CondA.Rep1_peaks.narrowPeak
condB_peaks=~/CourseData/EPI_data/module123/triplicates/peaks/CondB.Rep1_peaks.narrowPeak

cat ${condA_peaks} ${condB_peaks} | sort -k1,1 -k2,2n | bedtools merge -i stdin &gt; ~/workspace/module123/edgeR/merged_peaks.bed</code></pre>
<ul>
<li><code>cat ${condA_peaks} ${condB_peaks} | sort -k1,1 -k2,2n | bedtools merge -i stdin &gt; merged_peaks.bed</code>
<ul>
<li>Pseudo code break down : <code>Read our peaks | sort peaks coordinate wise | merge peaks</code></li>
</ul></li>
</ul>
</div>
<div id="step3b-differential-peaks-utilizing-fold-change-and-significance---merged-peaks" class="section level4 hasAnchor">
<h4>Step3B: Differential peaks utilizing Fold change and significance - Merged peaks<a href="module-3-chip-seq-differential-analysis.html#step3b-differential-peaks-utilizing-fold-change-and-significance---merged-peaks" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<ul>
<li>We’ll combine our two peak sets into an unified set.</li>
</ul>
<p><strong>Code:</strong></p>
<pre><code>###Shell###
condA_peaks=~/CourseData/EPI_data/module123/triplicates/peaks/CondA.Rep1_peaks.narrowPeak
condB_peaks=~/CourseData/EPI_data/module123/triplicates/peaks/CondB.Rep1_peaks.narrowPeak
mkdir ~/workspace/module123/edgeR/
cat ${condA_peaks} ${condB_peaks} | sort -k1,1 -k2,2n | bedtools merge -i stdin &gt; ~/workspace/module123/edgeR/merged_peaks.bed</code></pre>
<ul>
<li><code>cat ${condA_peaks} ${condB_peaks} | sort -k1,1 -k2,2n | bedtools merge -i stdin &gt; merged_peaks.bed</code>
<ul>
<li>Pseudo code break down : <code>Read our peaks | sort peaks coordinate wise | merge peaks</code></li>
</ul></li>
</ul>
</div>
<div id="step3b-differential-peaks-utilizing-fold-change-and-significance---read-counts-per-peak" class="section level4 hasAnchor">
<h4>Step3B: Differential peaks utilizing Fold change and significance - read counts per peak<a href="module-3-chip-seq-differential-analysis.html#step3b-differential-peaks-utilizing-fold-change-and-significance---read-counts-per-peak" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<ul>
<li>We’ll derrive RPKM values per BAM for each peak in our peak set</li>
</ul>
<p><strong>Code:</strong></p>
<pre><code>###Shell###
peaks=~/workspace/module123/edgeR/merged_peaks.bed
condA_bam=~/workspace/module123/alignments/MCF10A_H3K4me3_chr19.CondA.Rep1.bam
condB_bam=~/workspace/module123/alignments/MCF10A_H3K4me3_chr19.CondB.Rep1.bam

condA_count=~/workspace/module123/edgeR/MCF10A_H3K4me3_chr19.CondA.Rep1.bed
condB_count=~/workspace/module123/edgeR/MCF10A_H3K4me3_chr19.CondB.Rep1.bed

bedtools intersect -a ${peaks} -b ${condA_bam} -c &gt; ${condA_count}
bedtools intersect -a ${peaks} -b ${condB_bam} -c &gt; ${condB_count}</code></pre>
<ul>
<li><code>bedtools intersect -a ${peaks} -b ${condA_bam} -c &gt; ${condA_count}</code>
<ul>
<li>our intersect command is the same in principle but we’re intersecting our peaks with a <code>BAM</code> file</li>
<li><code>-c</code> reports counts of our <code>BAM</code> that overlap our peaks</li>
</ul></li>
</ul>
</div>
<div id="step3c-differential-peaks-utilizing-fold-change-and-significance---edger-differential" class="section level4 hasAnchor">
<h4>Step3C: Differential peaks utilizing Fold change and significance - EdgeR differential<a href="module-3-chip-seq-differential-analysis.html#step3c-differential-peaks-utilizing-fold-change-and-significance---edger-differential" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<ul>
<li>we’ll read our data into <code>R</code> and perform a statistical analysis</li>
</ul>
<p><strong>Code:</strong></p>
<pre><code></code></pre>
<p>###R###
setwd(“/home/ubuntu”)
library(edgeR)
library(dplyr)</p>
<p>condA&lt;-read.csv(“workspace/module123/edgeR/MCF10A_H3K4me3_chr19.CondA.Rep1.bed”,sep=‘,col.names = c(“chr”, “start”, “end”,“MCF10A_H3K4me3_chr19.CondA.Rep1”),colClasses= c(“character”,“character”,“character”,“numeric”))
condB&lt;-read.csv(“workspace/module123/edgeR/MCF10A_H3K4me3_chr19.CondB.Rep1.bed”,sep=’,col.names = c(“chr”, “start”, “end”,“MCF10A_H3K4me3_chr19.CondB.Rep1”),colClasses= c(“character”,“character”,“character”,“numeric”))</p>
<p>peaks&lt;-data.frame(
MCF10A_H3K4me3_chr19.CondA.Rep1=condA<span class="math inline">\(MCF10A_H3K4me3_chr19.CondA.Rep1,
    MCF10A_H3K4me3_chr19.CondB.Rep1=condB\)</span>MCF10A_H3K4me3_chr19.CondB.Rep1
)
row.names(peaks)&lt;-paste(condA<span class="math inline">\(chr,condA\)</span>start,condA$end,sep=’_’)</p>
<p>edger_dl &lt;- DGEList(counts=peaks, group=1:2,lib.size=c(1131503,1266436))</p>
<p>edger_tmm &lt;- calcNormFactors(edger_dl, method = “TMM”)</p>
<p>bvc=0.1</p>
<p>edger_et &lt;- exactTest(edger_tmm,dispersion=bvc^2)</p>
<p>edger_tp &lt;- topTags(edger_et, n=nrow(edger_et$table),adjust.method=“BH”)</p>
<p>de &lt;- edger_tp$table %&gt;% filter(FDR &lt; 0.01) %&gt;% filter(logFC &gt;=1 | logFC &lt;=-1)</p>
<p>write.table(de, file=“workspace/module123/edgeR/differential_peaks_edger.tsv”, sep=“, quote=F, row.names=F, col.names=F)</p>
<pre><code>- `condA&lt;-read.csv(&quot;workspace/module123/edgeR/MCF10A_H3K4me3_chr19.CondA.Rep1.bed&quot;,sep=&#39;\t&#39;,col.names = c(&quot;chr&quot;, &quot;start&quot;, &quot;end&quot;,&quot;MCF10A_H3K4me3_chr19.CondA.Rep1&quot;),colClasses= c(&quot;character&quot;,&quot;character&quot;,&quot;character&quot;,&quot;numeric&quot;))`
    - `read.csv` read the `CSV` file into a `data.frame`
    - `sep=&#39;\t&#39;` specify the delimiter
    - `col.names = c(&quot;chr&quot;, &quot;start&quot;, &quot;end&quot;,&quot;MCF10A_H3K4me3_chr19.CondA.Rep1&quot;)` set our column names
    - `colClasses= c(&quot;character&quot;,&quot;character&quot;,&quot;character&quot;,&quot;numeric&quot;)` indicate with column are what datatypes
- `peaks&lt;-data.frame(MCF10A_H3K4me3_chr19.CondA.Rep1=condA$MCF10A_H3K4me3_chr19.CondA.Rep1,MCF10A_H3K4me3_chr19.CondB.Rep1=condB$MCF10A_H3K4me3_chr19.CondB.Rep1)`
    - make a new data.frame using our two columns from condA and condB
- `row.names(peaks)&lt;-paste(peaks$chr,peaks$start,peaks$end,sep=&#39;_&#39;)`
    - edit row names to be match genomic coordinates
    - ``row.names(peaks)&lt;-` indicate we want to overwrite exist row names
    - `paste(peaks$chr,peaks$start,peaks$end,sep=&#39;_&#39;)` we want to concatenate our `chr`,`start` and `end` with `_` as a seperator
- `edger_dl &lt;- DGEList(counts=peaks, group=1:2,lib.size=c(1131503,1266436))`
    - `DGEList(counts=peaks_clean, group=1:2)` read in our `peak_clean` data.frame
    - `group=1:2` identify which groups we want to contrast
    - `lib.size=c(1131503,1266436)` library size for the two conditions
- `edger_tmm &lt;- calcNormFactors(edger_dl, method = &quot;TMM&quot;)` calculate the normalization factor to scale library sizes
- `bvc=0.01` Set the `square-rootdispersion`. according to edgeR documentation: `from well-controlled experiments are 0.4 for human data, 0.1 for data on genetically identical model organisms or 0.01 for technical replicates`
- `edger_et &lt;- exactTest(edger_dl,dispersion=bcv^2)`
    - calculate `FC` and `Pvalue` per row
- `edger_tp &lt;- topTags(edger_et, n=nrow(edger_et$table),adjust.method=&quot;BH&quot;)`
    - calcualte `FDR` via Benjamini-Hochberg
- `de &lt;- edger_tp$table %&gt;% filter(FDR &lt; 0.01) %&gt;% filter(logFC &gt;=1 | logFC &lt;=-1)`
    - `filter(FDR &lt; 0.01)` filter for significant peaks
    - `filter(logFC &gt;=1 | logFC &lt;=-1)` filter for peaks with appropriate fold change
- `write.table(de, file=&quot;workspace/module123/edgeR/differential_peaks_edger.tsv&quot;, sep=&quot;\t&quot;, quote=F, row.names=F, col.names=F)`
    - save files
&lt;img src=&quot;https://github.com/bioinformaticsdotca/EPI_2023/blob/module123/module123_images/edger.png?raw=true&quot; alt=&quot;Region&quot; width=&quot;750&quot; /&gt;</code></pre>
</div>
</div>
<div id="server-resources-2" class="section level3 hasAnchor">
<h3>Server resources<a href="module-3-chip-seq-differential-analysis.html#server-resources-2" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<div id="qc-resources-2" class="section level4 hasAnchor">
<h4>QC Resources<a href="module-3-chip-seq-differential-analysis.html#qc-resources-2" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<pre><code>###TSS+/-2kb
mkdir workspace/module123/qc
wget https://www.bcgsc.ca/downloads/esu/touchdown/hg38v79_genes_tss_2000.bed -O workspace/module123/resources/hg38v79_genes_tss_2000.bed

sort -k1,1 -k2,2n workspace/module123/resources/hg38v79_genes_tss_2000.bed &gt; tmp
mv tmp workspace/module123/resources/hg38v79_genes_tss_2000.bed

###Enhancer liftover
wget https://www.bcgsc.ca/downloads/esu/touchdown/encode_enhancers_liftover.bed -O workspace/module123/resources/encode_enhancers_liftover.bed

###Blacklist
wget https://www.encodeproject.org/files/ENCFF356LFX/@@download/ENCFF356LFX.bed.gz -O ~/workspace/module123/resources/hg38_blacklist.bed.gz

gunzip ~/workspace/module123/resources/hg38_blacklist.bed.gz</code></pre>
<ul>
<li><code>hg38v79_genes_tss_2000.bed</code>
<ul>
<li>Generated by downloading Ensemblv79 GTF convert to Bed +/-2kb of TSS. See <a href="https://www.biostars.org/p/56280/" class="uri">https://www.biostars.org/p/56280/</a></li>
</ul></li>
<li><code>encode_enhancers_liftover.bed</code>
<ul>
<li>download various <a href="https://egg2.wustl.edu/roadmap/data/byFileType/chromhmmSegmentations/ChmmModels/coreMarks/jointModel/final/">ChroHMM state7</a> and merge</li>
</ul></li>
</ul>
</div>
<div id="encode-bed-2" class="section level4 hasAnchor">
<h4>Encode Bed<a href="module-3-chip-seq-differential-analysis.html#encode-bed-2" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<pre><code>ls ~/CourseData/EPI_data/module123/encode_bed
basal.H3K27ac.peak_calls.bed
basal.H3K27me3.peak_calls.bed
basal.H3K4me1.peak_calls.bed
basal.H3K4me3.peak_calls.bed
lp.H3K27ac.peak_calls.bed
lp.H3K27me3.peak_calls.bed
lp.H3K4me1.peak_calls.bed
lp.H3K4me3.peak_calls.bed
luminal.H3K27ac.peak_calls.bed
luminal.H3K27me3.peak_calls.bed
luminal.H3K4me1.peak_calls.bed
luminal.H3K4me3.peak_calls.bed
stromal.H3K27ac.peak_calls.bed
stromal.H3K27me3.peak_calls.bed
stromal.H3K4me1.peak_calls.bed
stromal.H3K4me3.peak_calls.bed</code></pre>
<ul>
<li><a href="https://epigenomesportal.ca/tracks/CEEHRC/hg38/" class="uri">https://epigenomesportal.ca/tracks/CEEHRC/hg38/</a></li>
<li>Breast Basal CEMT0035</li>
<li>Breast Stromal CEMT0036</li>
<li>Breast Luminal CEMT0037</li>
<li>Breast Luminal Progenitor CEMT0038
#### Encode BigWig</li>
</ul>
<pre><code>ls ~/CourseData/EPI_data/module123/encode_bigWig
basal.H3K27ac.signal_unstranded.bigWig
basal.H3K27me3.signal_unstranded.bigWig
basal.H3K4me1.signal_unstranded.bigWig
basal.H3K4me3.signal_unstranded.bigWig
lp.H3K27ac.signal_unstranded.bigWig
lp.H3K27me3.signal_unstranded.bigWig
lp.H3K4me1.signal_unstranded.bigWig
lp.H3K4me3.signal_unstranded.bigWig
luminal.H3K27ac.signal_unstranded.bigWig
luminal.H3K27me3.signal_unstranded.bigWig
luminal.H3K4me1.signal_unstranded.bigWig
luminal.H3K4me3.signal_unstranded.bigWig
stromal.H3K27ac.signal_unstranded.bigWig
stromal.H3K27me3.signal_unstranded.bigWig
stromal.H3K4me1.signal_unstranded.bigWig
stromal.H3K4me3.signal_unstranded.bigWig</code></pre>
<ul>
<li><a href="https://epigenomesportal.ca/tracks/CEEHRC/hg38/" class="uri">https://epigenomesportal.ca/tracks/CEEHRC/hg38/</a></li>
<li>Breast Basal CEMT0035</li>
<li>Breast Stromal CEMT0036</li>
<li>Breast Luminal CEMT0037</li>
<li>Breast Luminal Progenitor CEMT0038
#### MCF10A Fastq</li>
</ul>
<pre><code>ls ~/CourseData/EPI_data/module123/fastq
MCF10A.ATAC.chr19.R1.fastq.gz
MCF10A.ATAC.chr19.R2.fastq.gz
MCF10A.H3K27ac.chr19.R1.fastq.gz
MCF10A.H3K27ac.chr19.R2.fastq.gz
MCF10A.H3K27me3.chr19.R1.fastq.gz
MCF10A.H3K27me3.chr19.R2.fastq.gz
MCF10A.H3K4me3.chr19.R1.fastq.gz
MCF10A.H3K4me3.chr19.R2.fastq.gz
MCF10A.Input.chr19.R1.fastq.gz
MCF10A.Input.chr19.R2.fastq.gz</code></pre>
<ul>
<li>MCF10A histone marks and input come courtesy of Dr.Hirst</li>
<li>ATACseq data originates from <a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSM6431322">GSM6431322</a>
#### Triplicates</li>
</ul>
<pre><code>CourseData/EPI_data/module123/triplicates/triplicates.csv

CourseData/EPI_data/module123/triplicates/alignments:
MCF10A_H3K4me3_chr19.CondA.Rep1.bam      MCF10A_H3K4me3_chr19.CondB.Rep2.bam      MCF10A_input_chr19.CondA.Rep3.bam
MCF10A_H3K4me3_chr19.CondA.Rep1.bam.bai  MCF10A_H3K4me3_chr19.CondB.Rep2.bam.bai  MCF10A_input_chr19.CondA.Rep3.bam.bai
MCF10A_H3K4me3_chr19.CondA.Rep2.bam      MCF10A_H3K4me3_chr19.CondB.Rep3.bam      MCF10A_input_chr19.CondB.Rep1.bam
MCF10A_H3K4me3_chr19.CondA.Rep2.bam.bai  MCF10A_H3K4me3_chr19.CondB.Rep3.bam.bai  MCF10A_input_chr19.CondB.Rep1.bam.bai
MCF10A_H3K4me3_chr19.CondA.Rep3.bam      MCF10A_input_chr19.CondA.Rep1.bam        MCF10A_input_chr19.CondB.Rep2.bam
MCF10A_H3K4me3_chr19.CondA.Rep3.bam.bai  MCF10A_input_chr19.CondA.Rep1.bam.bai    MCF10A_input_chr19.CondB.Rep2.bam.bai
MCF10A_H3K4me3_chr19.CondB.Rep1.bam      MCF10A_input_chr19.CondA.Rep2.bam        MCF10A_input_chr19.CondB.Rep3.bam
MCF10A_H3K4me3_chr19.CondB.Rep1.bam.bai  MCF10A_input_chr19.CondA.Rep2.bam.bai    MCF10A_input_chr19.CondB.Rep3.bam.bai

CourseData/EPI_data/module123/triplicates/bigWig:
CondA.Rep1.bw  CondA.Rep2.bw  CondA.Rep3.bw  CondB.Rep1.bw  CondB.Rep2.bw  CondB.Rep3.bw

CourseData/EPI_data/module123/triplicates/peaks:
CondA.Rep1_peaks.narrowPeak  CondA.Rep3_peaks.narrowPeak  CondB.Rep2_peaks.narrowPeak
CondA.Rep2_peaks.narrowPeak  CondB.Rep1_peaks.narrowPeak  CondB.Rep3_peaks.narrowPeak</code></pre>
<ul>
<li>triplicates were generated from MCF10A_H3K4me3 by choosing a list of exclusive peaks for condA and condB and subsampling replicates accordingly</li>
</ul>
<div class="callout-green">
<span class="callout-icon"><i class="fas fa-circle-check" aria-hidden="true"></i></span>
<div class="callout-content">
<div class="callout-title-static">
<p>Lab Completed!</p>
</div>
<p>Congratulations! You have completed Lab 3!</p>
</div>
</div>

</div>
</div>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="module-2-alignment-peak-calling-and-visualization.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="module-4-whole-genome-bisulfite-sequencing-and-analysis.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
  "sharing": {
    "github": false,
    "facebook": false,
    "twitter": false,
    "linkedin": false,
    "weibo": false,
    "instapaper": false,
    "vk": false,
    "whatsapp": false,
    "all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
  },
  "fontsettings": {
    "theme": "white",
    "family": "sans",
    "size": 2
  },
  "edit": {
    "link": "https://github.com/bioinformaticsdotca/EPI_2023/blob/main/030-module-3.Rmd",
    "text": "Edit"
  },
  "history": {
    "link": null,
    "text": null
  },
  "view": {
    "link": null,
    "text": null
  },
  "download": ["_main.pdf", "_main.epub"],
  "search": {
    "engine": "fuse",
    "options": null
  },
  "toc": {
    "collapse": "subsection"
  }
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.9/latest.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
