<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Module 2: Alignment, Peak Calling, and Visualization | Epigenomic Analysis 2023</title>
  <meta name="description" content="Workshop website for the {{YEAR}} {{WORKSHOP_NAME}} Canadian Bioinformatics Workshop" />
  <meta name="generator" content="bookdown 0.43 and GitBook 2.6.7" />

  <meta property="og:title" content="Module 2: Alignment, Peak Calling, and Visualization | Epigenomic Analysis 2023" />
  <meta property="og:type" content="book" />
  <meta property="og:image" content="https://bioinformaticsdotca.github.io/EPI_2023/img/bioinformatics_logo.png" />
  <meta property="og:description" content="Workshop website for the {{YEAR}} {{WORKSHOP_NAME}} Canadian Bioinformatics Workshop" />
  <meta name="github-repo" content="bioinformaticsdotca/EPI_2023" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Module 2: Alignment, Peak Calling, and Visualization | Epigenomic Analysis 2023" />
  
  <meta name="twitter:description" content="Workshop website for the {{YEAR}} {{WORKSHOP_NAME}} Canadian Bioinformatics Workshop" />
  <meta name="twitter:image" content="https://bioinformaticsdotca.github.io/EPI_2023/img/bioinformatics_logo.png" />

<meta name="author" content="Faculty: Guillaume Bourque, Martin Hirst, David Bujold, Jose Hector Galvez, Edmund Su, Mareike Janiak, Michelle Brazas, Nia Hughes, and Zhibin Lu" />



  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  <link rel="shortcut icon" href="img/favicon.ico" type="image/x-icon" />
<link rel="prev" href="module-1-introduction-to-chip-sequencing-and-analysis.html"/>
<link rel="next" href="module-3-chip-seq-differential-analysis.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.5.1/css/all.min.css">
<script type="application/ld+json">
{
  "@context": "https://schema.org",
  "@type": "LearningResource",
  "name": "Epigenomic Analysis",
  "description": "The complete set of tutorials, slides, and practical exercises for the Epigenomic Analysis 2023 Canadian Bioinformatics Workshop, hosted on GitHub Pages.",
  "url": "https://bioinformaticsdotca.github.io/EPI_2023",
  "learningResourceType": ["tutorial", "slides", "exercise"],
  "inLanguage": "en",
  "license": "https://creativecommons.org/licenses/by/4.0/",
  "provider": {
    "@type": "Organization",
    "name": "Bioinformatics.ca",
    "url": "https://bioinformatics.ca"
  },
  "keywords": "",
  "isPartOf": {
    "@type": "Course",
    "@id": "http://bioinformaticsdotca.github.io/EPI_2023"
  }
}
</script>


<style type="text/css">
html { -webkit-text-size-adjust: 100%; }
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
  
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<a href="https://bioinformaticsdotca.github.io/">
 <img src="img/bioinformatics_logo.png" width="90%" alt="bioinformatics.ca logo" style="display:block; margin-left:auto; margin-right:auto; margin-top:15px;">
 </a>
<li><center><strong><a href="./"><br>Epigenomic Analysis<br><br></a></strong></center></li>
<a href="./">
 <img src="img/CBW_Epigenome-data_icon.png" width="70%" alt="Epigenomic Analysis logo" style="display:block; margin-left:auto; margin-right:auto; ">
 </a>
<br>

<li class="divider"></li>
<li class="part"><span><b>I Introduction</b></span></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Workshop Info</a>
<ul>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#pre-work"><i class="fa fa-check"></i>Pre-work</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#class-photo"><i class="fa fa-check"></i>Class Photo</a></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html#schedule"><i class="fa fa-check"></i>Schedule</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="meet-your-faculty.html"><a href="meet-your-faculty.html"><i class="fa fa-check"></i>Meet Your Faculty</a></li>
<li class="chapter" data-level="" data-path="data-and-compute-setup.html"><a href="data-and-compute-setup.html"><i class="fa fa-check"></i>Data and Compute Setup</a></li>
<li class="part"><span><b>II Modules</b></span></li>
<li class="chapter" data-level="" data-path="module-1-introduction-to-chip-sequencing-and-analysis.html"><a href="module-1-introduction-to-chip-sequencing-and-analysis.html"><i class="fa fa-check"></i>Module 1: Introduction to ChIP Sequencing and Analysis</a>
<ul>
<li class="chapter" data-level="" data-path="module-1-introduction-to-chip-sequencing-and-analysis.html"><a href="module-1-introduction-to-chip-sequencing-and-analysis.html#lecture"><i class="fa fa-check"></i>Lecture</a></li>
<li class="chapter" data-level="" data-path="module-1-introduction-to-chip-sequencing-and-analysis.html"><a href="module-1-introduction-to-chip-sequencing-and-analysis.html#lab"><i class="fa fa-check"></i>Lab</a>
<ul>
<li class="chapter" data-level="" data-path="module-1-introduction-to-chip-sequencing-and-analysis.html"><a href="module-1-introduction-to-chip-sequencing-and-analysis.html#breakdown-of-markdown-file"><i class="fa fa-check"></i>Breakdown of Markdown File</a></li>
<li class="chapter" data-level="" data-path="module-1-introduction-to-chip-sequencing-and-analysis.html"><a href="module-1-introduction-to-chip-sequencing-and-analysis.html#module-1.-alignment"><i class="fa fa-check"></i>Module 1. Alignment</a></li>
<li class="chapter" data-level="" data-path="module-1-introduction-to-chip-sequencing-and-analysis.html"><a href="module-1-introduction-to-chip-sequencing-and-analysis.html#server-resources"><i class="fa fa-check"></i>Server resources</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="" data-path="module-2-alignment-peak-calling-and-visualization.html"><a href="module-2-alignment-peak-calling-and-visualization.html"><i class="fa fa-check"></i>Module 2: Alignment, Peak Calling, and Visualization</a>
<ul>
<li class="chapter" data-level="" data-path="module-2-alignment-peak-calling-and-visualization.html"><a href="module-2-alignment-peak-calling-and-visualization.html#lecture-1"><i class="fa fa-check"></i>Lecture</a></li>
<li class="chapter" data-level="" data-path="module-2-alignment-peak-calling-and-visualization.html"><a href="module-2-alignment-peak-calling-and-visualization.html#lab-1"><i class="fa fa-check"></i>Lab</a>
<ul>
<li class="chapter" data-level="" data-path="module-2-alignment-peak-calling-and-visualization.html"><a href="module-2-alignment-peak-calling-and-visualization.html#breakdown-of-markdown-file-1"><i class="fa fa-check"></i>Breakdown of Markdown File</a></li>
<li class="chapter" data-level="" data-path="module-2-alignment-peak-calling-and-visualization.html"><a href="module-2-alignment-peak-calling-and-visualization.html#module-2.-peak-calling"><i class="fa fa-check"></i>Module 2. Peak Calling</a></li>
<li class="chapter" data-level="" data-path="module-2-alignment-peak-calling-and-visualization.html"><a href="module-2-alignment-peak-calling-and-visualization.html#server-resources-1"><i class="fa fa-check"></i>Server resources</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="" data-path="module-3-chip-seq-differential-analysis.html"><a href="module-3-chip-seq-differential-analysis.html"><i class="fa fa-check"></i>Module 3: ChIP-seq Differential Analysis</a>
<ul>
<li class="chapter" data-level="" data-path="module-3-chip-seq-differential-analysis.html"><a href="module-3-chip-seq-differential-analysis.html#lecture-2"><i class="fa fa-check"></i>Lecture</a></li>
<li class="chapter" data-level="" data-path="module-3-chip-seq-differential-analysis.html"><a href="module-3-chip-seq-differential-analysis.html#lab-2"><i class="fa fa-check"></i>Lab</a>
<ul>
<li class="chapter" data-level="" data-path="module-3-chip-seq-differential-analysis.html"><a href="module-3-chip-seq-differential-analysis.html#breakdown-of-markdown-file-2"><i class="fa fa-check"></i>Breakdown of markdown file</a></li>
<li class="chapter" data-level="" data-path="module-3-chip-seq-differential-analysis.html"><a href="module-3-chip-seq-differential-analysis.html#module-3---differential-analysis"><i class="fa fa-check"></i>Module 3 - Differential Analysis</a></li>
<li class="chapter" data-level="" data-path="module-3-chip-seq-differential-analysis.html"><a href="module-3-chip-seq-differential-analysis.html#server-resources-2"><i class="fa fa-check"></i>Server resources</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="" data-path="module-4-whole-genome-bisulfite-sequencing-and-analysis.html"><a href="module-4-whole-genome-bisulfite-sequencing-and-analysis.html"><i class="fa fa-check"></i>Module 4: Whole-Genome Bisulfite Sequencing and Analysis</a>
<ul>
<li class="chapter" data-level="" data-path="module-4-whole-genome-bisulfite-sequencing-and-analysis.html"><a href="module-4-whole-genome-bisulfite-sequencing-and-analysis.html#lecture-3"><i class="fa fa-check"></i>Lecture</a></li>
<li class="chapter" data-level="" data-path="module-4-whole-genome-bisulfite-sequencing-and-analysis.html"><a href="module-4-whole-genome-bisulfite-sequencing-and-analysis.html#lab-3"><i class="fa fa-check"></i>Lab</a>
<ul>
<li class="chapter" data-level="" data-path="module-4-whole-genome-bisulfite-sequencing-and-analysis.html"><a href="module-4-whole-genome-bisulfite-sequencing-and-analysis.html#introduction"><i class="fa fa-check"></i>1. Introduction</a></li>
<li class="chapter" data-level="" data-path="module-4-whole-genome-bisulfite-sequencing-and-analysis.html"><a href="module-4-whole-genome-bisulfite-sequencing-and-analysis.html#mapping-tutorial"><i class="fa fa-check"></i>2. Mapping Tutorial</a></li>
<li class="chapter" data-level="" data-path="module-4-whole-genome-bisulfite-sequencing-and-analysis.html"><a href="module-4-whole-genome-bisulfite-sequencing-and-analysis.html#differential-methylation-analysis-in-methylkit"><i class="fa fa-check"></i>3. Differential Methylation Analysis in MethylKit</a></li>
<li class="chapter" data-level="" data-path="module-4-whole-genome-bisulfite-sequencing-and-analysis.html"><a href="module-4-whole-genome-bisulfite-sequencing-and-analysis.html#differential-methylation-analysis-in-dss"><i class="fa fa-check"></i>4. Differential Methylation Analysis in DSS</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="" data-path="module-5-downstream-analysis-and-online-tools.html"><a href="module-5-downstream-analysis-and-online-tools.html"><i class="fa fa-check"></i>Module 5: Downstream Analysis and Online Tools</a>
<ul>
<li class="chapter" data-level="" data-path="module-5-downstream-analysis-and-online-tools.html"><a href="module-5-downstream-analysis-and-online-tools.html#lecture-4"><i class="fa fa-check"></i>Lecture</a></li>
<li class="chapter" data-level="" data-path="module-5-downstream-analysis-and-online-tools.html"><a href="module-5-downstream-analysis-and-online-tools.html#lab-4"><i class="fa fa-check"></i>Lab</a>
<ul>
<li class="chapter" data-level="" data-path="module-5-downstream-analysis-and-online-tools.html"><a href="module-5-downstream-analysis-and-online-tools.html#introduction-1"><i class="fa fa-check"></i>Introduction</a></li>
<li class="chapter" data-level="" data-path="module-5-downstream-analysis-and-online-tools.html"><a href="module-5-downstream-analysis-and-online-tools.html#ihec-data-portal"><i class="fa fa-check"></i>IHEC Data Portal</a></li>
<li class="chapter" data-level="" data-path="module-5-downstream-analysis-and-online-tools.html"><a href="module-5-downstream-analysis-and-online-tools.html#predicting-motifs-with-homer"><i class="fa fa-check"></i>Predicting Motifs with HOMER</a></li>
<li class="chapter" data-level="" data-path="module-5-downstream-analysis-and-online-tools.html"><a href="module-5-downstream-analysis-and-online-tools.html#looking-for-go-terms-enrichment-with-great"><i class="fa fa-check"></i>Looking for GO Terms Enrichment with GREAT</a></li>
<li class="chapter" data-level="" data-path="module-5-downstream-analysis-and-online-tools.html"><a href="module-5-downstream-analysis-and-online-tools.html#going-back-to-homer-results"><i class="fa fa-check"></i>Going back to HOMER Results</a></li>
</ul></li>
</ul></li>
<li class="divider"></li>
<h1 id="sponsors">Sponsors</h1>
<center>
<<a href="https://genomecanada.ca/"><img src="img/sponsors/genomecanada.png" height="70" style="display:inline-block;"  style="padding-bottom: 20px" alt="Genome Canada logo.">
<a href="https://www.ontariogenomics.ca/"><img src="img/sponsors/ontariogenomics.png" height="70" style="display:inline-block;" style="padding-bottom: 20px;" alt="Ontario Genomics logo.">
<br><br>
<a href="https://oicr.on.ca/"><img src="img/sponsors/oicr.png" width=25% style="display:inline-block;" style="padding-bottom: 20px" alt="Ontario Institute for Cancer Research logo.">
<a href="https://www.hpcforhealth.ca/"><img src="img/sponsors/hpc4health.png" width=60% style="display:inline-block;" style="padding-bottom: 20px "alt="HPC4Health logo.">
<br><br>
<a href="https://computationalgenomics.ca/"><img src="img/sponsors/c3g_Web.png" style="display:inline-block;" width=35% style="padding-bottom: 20px" alt="Canadian Centre for Computational Genomcis logo.">

<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li></center>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Epigenomic Analysis 2023</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="module-2-alignment-peak-calling-and-visualization" class="section level1 hasAnchor">
<h1>Module 2: Alignment, Peak Calling, and Visualization<a href="module-2-alignment-peak-calling-and-visualization.html#module-2-alignment-peak-calling-and-visualization" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<div id="lecture-1" class="section level2 hasAnchor">
<h2>Lecture<a href="module-2-alignment-peak-calling-and-visualization.html#lecture-1" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<iframe width="640" height="360" src="https://www.youtube.com/embed/MrxiepiBICQ?si=ZZ-HuxMFoC7d1TGg" title="YouTube video player" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture; web-share" referrerpolicy="strict-origin-when-cross-origin" allowfullscreen>
</iframe>
<p><br></p>
<iframe src="https://drive.google.com/file/d/1_qP_eCe_nXooe8p1DKLgpWGpjsagAV-q/preview" width="640" height="480" allow="autoplay">
</iframe>
</div>
<div id="lab-1" class="section level2 hasAnchor">
<h2>Lab<a href="module-2-alignment-peak-calling-and-visualization.html#lab-1" class="anchor-section" aria-label="Anchor link to header"></a></h2>
<div id="breakdown-of-markdown-file-1" class="section level3 hasAnchor">
<h3>Breakdown of Markdown File<a href="module-2-alignment-peak-calling-and-visualization.html#breakdown-of-markdown-file-1" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<ul>
<li>At the start of each step, the intention will declare.</li>
<li>this is then follow by a code block</li>
</ul>
<p><strong>Code:</strong></p>
<pre><code>Like this! This is the main code to run for the step.

Additionally the code block will include a header to indicate what environment to the run code for example:
###Shell###
pwd

###R###
getwd()</code></pre>
<ul>
<li>explaining for commands will be broken down following code block</li>
<li><code>pwd</code> &amp; <code>getwd()</code>
<ul>
<li>see your work directory</li>
</ul></li>
<li>sprinkled throughout will also include comments</li>
</ul>
<div class="callout-blue">
<span class="callout-icon"><i class="fas fa-circle-info" aria-hidden="true"></i></span>
<div class="callout-content">
<div class="callout-title-static">
<p>Note</p>
</div>
<p>Important points and considerations will also be raised as so.</p>
</div>
</div>
</div>
<div id="module-2.-peak-calling" class="section level3 hasAnchor">
<h3>Module 2. Peak Calling<a href="module-2-alignment-peak-calling-and-visualization.html#module-2.-peak-calling" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<div id="step-1-dedup-bam-file" class="section level4 hasAnchor">
<h4>Step 1: Dedup BAM file<a href="module-2-alignment-peak-calling-and-visualization.html#step-1-dedup-bam-file" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<ul>
<li>We first remove duplicates here b/c <code>MACS2</code>(peak caller) identifies duplicates via genomic coordinates (excessive removal of a lot reads).</li>
<li>dedup now and instruct our peak caller to “keep duplicates” after.</li>
</ul>
<p><strong>Code:</strong></p>
<pre><code>###Shell###
treatment=MCF10A_H3K27ac_chr19
treatment_bam=~/workspace/module123/alignments/${treatment}.sorted.dup_marked.bam
treatment_dedup=~/workspace/module123/alignments/${treatment}.sorted.dup_marked.dedup.bam
samtools view -@4 ${treatment_bam} -bh -q10 -F1028 -o ${treatment_dedup}

treatment=MCF10A_H3K27me3_chr19
treatment_bam=~/workspace/module123/alignments/${treatment}.sorted.dup_marked.bam
treatment_dedup=~/workspace/module123/alignments/${treatment}.sorted.dup_marked.dedup.bam
samtools view -@4 ${treatment_bam} -bh -q10 -F1028 -o ${treatment_dedup}

treatment=MCF10A_H3K4me3_chr19
treatment_bam=~/workspace/module123/alignments/${treatment}.sorted.dup_marked.bam
treatment_dedup=~/workspace/module123/alignments/${treatment}.sorted.dup_marked.dedup.bam
samtools view -@4 ${treatment_bam} -bh -q10 -F1028 -o ${treatment_dedup}

treatment=MCF10A_ATAC_chr19
treatment_bam=~/workspace/module123/alignments/${treatment}.sorted.dup_marked.bam
treatment_dedup=~/workspace/module123/alignments/${treatment}.sorted.dup_marked.dedup.bam
samtools view -@4 ${treatment_bam} -bh -q10 -F1028 -o ${treatment_dedup}

input=MCF10A_input_chr19
input_bam=~/workspace/module123/alignments/${input}.sorted.dup_marked.bam
input_dedup=~/workspace/module123/alignments/${input}.sorted.dup_marked.dedup.bam
samtools view -@4 ${input_bam} -bh -q10 -F1028 -o ${input_dedup}</code></pre>
<ul>
<li>Run time &lt;1 min per command</li>
<li>Silently completes; no log needed</li>
<li><code>samtools view -@4 ${input_bam} -bh -q10 -F1028 -o ${input_dedup}</code>
<ul>
<li>we subset our file based on the following criteria</li>
<li><code>-bh</code> we would like the output to be in <code>BAM</code> format and contain the header</li>
<li><code>-q10</code> reads with mapping qual <code>&gt;10</code></li>
<li><code>-F1028</code> any reads that do not have the flags <code>UNMAP</code> and <code>DUP</code>
<ul>
<li>whats the difference between <code>-f</code> and <code>-F</code>?
### Step 2A : Run Peak Caller (narrow)</li>
</ul></li>
</ul></li>
<li>MACS has two modes for narrow marks and broad marks.</li>
<li>Refer to this <a href="https://www.encodeproject.org/chip-seq/histone/#:~:text=must%20pass%20routine%20metadata%20audits%20in%20order%20to%20be%20released.-,Target%2Dspecific%20Standards,-For%20narrow%2Dpeak%20histone%20experiments%2C%20each%20replicate%C2%A0should%20have%2020">link</a> to reference which mark are narrow or broad</li>
</ul>
<p><strong>Code:</strong></p>
<pre><code>###Shell###
mkdir -p ~/workspace/module123/peaks
name=MCF10A_H3K27ac
treatment=~/workspace/module123/alignments/${name}_chr19.sorted.dup_marked.dedup.bam
input=~/workspace/module123/alignments/MCF10A_input_chr19.sorted.dup_marked.dedup.bam


macs2 callpeak -t ${treatment} -c ${input} -f BAMPE -g 58617616 -n ${name} --keep-dup all --outdir ~/workspace/module123/peaks/ --bdg 1&gt; ~/workspace/module123/peaks/${name}.out.log 2&gt; ~/workspace/module123/peaks/${name}.err.log

name=MCF10A_H3K4me3
treatment=~/workspace/module123/alignments/${name}_chr19.sorted.dup_marked.dedup.bam
input=~/workspace/module123/alignments/MCF10A_input_chr19.sorted.dup_marked.dedup.bam


macs2 callpeak -t ${treatment} -c ${input} -f BAMPE -g 58617616 -n ${name} --keep-dup all --outdir ~/workspace/module123/peaks/ --bdg 1&gt; ~/workspace/module123/peaks/${name}.out.log 2&gt; ~/workspace/module123/peaks/${name}.err.log</code></pre>
<ul>
<li><code>macs2 callpeak -t ${treatment} -c ${input} -f BAMPE -g 58617616 -n ${name} --keep-dup all --outdir ~/workspace/module123/peaks/ --bdg 1&gt; ~/workspace/module123/peaks/${name}.out.log 2&gt; ~/workspace/module123/peaks/${name}.err.log</code>
<ul>
<li><code>macs2 callpeak</code> General purpose peak calling mode</li>
<li><code>-t ${treatment}</code> Treatment file, can provide more than one to “pool”</li>
<li><code>-c ${input}</code> Control File/Input</li>
<li><code>-f BAMPE</code> Instructs MACS2 on what kind of file to expect. Single/Paired-end bed/bam</li>
<li><code>-g hs</code> Sets appropriate genome size for background sampling. Typically would set <code>hs=human</code> <code>mm=mouse</code> but in our case we use the HG38 size of chr19</li>
<li><code>-n ${name}</code> name or prefix to use</li>
<li><code>-q 0.05</code> FDR q value default</li>
<li><code>--outdir ~/workspace/module123/peaks/</code> where to output files otherwise stores in current working directory</li>
<li><code>--bdg</code> outputs pileup into bedgraph (a <code>BED</code> file where the fourth column is pileup/fragment count/coverage)</li>
<li><code>1&gt; ~/workspace/module123/peaks/${name}.out.log</code> output log</li>
<li><code>2&gt;  ~/workspace/module123/peaks/${name}.err.log</code> error log</li>
<li>let’s inspect the peaks file
<ul>
<li>chromosome name</li>
<li>peak start</li>
<li>peak stop</li>
<li>peak name</li>
<li>int(-10*log10pvalue)</li>
<li>strand</li>
<li>Fold change at peak summit</li>
<li>-log10 P-value at Peak</li>
<li>-log10 Q-value at Peak</li>
<li>Summit position relative to peak</li>
</ul></li>
</ul></li>
</ul>
</div>
<div id="step-2b-run-peak-caller-broad" class="section level4 hasAnchor">
<h4>Step 2B : Run Peak Caller (broad)<a href="module-2-alignment-peak-calling-and-visualization.html#step-2b-run-peak-caller-broad" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p><strong>Code:</strong></p>
<pre><code>###Shell###
mkdir -p ~/workspace/module123/peaks
name=MCF10A_H3K27me3
treatment=~/workspace/module123/alignments/${name}_chr19.sorted.dup_marked.dedup.bam
input=~/workspace/module123/alignments/MCF10A_input_chr19.sorted.dup_marked.dedup.bam

macs2 callpeak -t ${treatment} -c ${input} -f BAMPE -g 58617616 -n ${name} --keep-dup all --outdir ~/workspace/module123/peaks/ --bdg --broad 1&gt; ~/workspace/module123/peaks/${name}.out.log 2&gt; ~/workspace/module123/peaks/${name}.err.log</code></pre>
<ul>
<li><code>macs2 callpeak -t ${treatment} -c ${input} -f BAMPE -g 58617616 -n ${name} --keep-dup all --outdir ~/workspace/module123/peaks/ --bdg --broad 1&gt; ~/workspace/module123/peaks/${name}.out.log 2&gt; ~/workspace/module123/peaks/${name}.err.log</code>
<ul>
<li><code>--broad</code> for broad marks - stitches small peaks together</li>
<li>let’s note the differences between <code>broadPeak</code> vs <code>gappedPeak</code></li>
<li><code>gapped</code> has the additional columns starting from <code>strand</code> where most of the differences are visualization support for the narrrow peaks within the broadPeak
<ul>
<li>thickStart</li>
<li>thickEnd</li>
<li>itemRgb</li>
<li>blockCount</li>
<li>blockSizes</li>
<li>blockStarts</li>
<li>signalValue</li>
<li>pValue</li>
<li>qValue
### Step 3A : Run Peak Caller for ATAC - Make fragment file</li>
</ul></li>
</ul></li>
<li>Convert our <code>BAM</code> to <code>BED</code> to easily manipulate coordinates</li>
<li>perform shift to account for Tn5 binding as a dimer and inserts two adapters separated by 9 bp</li>
</ul>
<div class="callout-blue">
<span class="callout-icon"><i class="fas fa-circle-info" aria-hidden="true"></i></span>
<div class="callout-content">
<div class="callout-title-static">
<p>Note</p>
</div>
<p>This step can also be done for chipseq</p>
</div>
</div>
<p><strong>Code:</strong></p>
<pre><code>name=MCF10A_ATAC
dedup=~/workspace/module123/alignments/${name}_chr19.sorted.dup_marked.dedup.bam
nsort=~/workspace/module123/alignments/${name}_chr19.nsorted.dup_marked.dedup.bam
tags=~/workspace/module123/peaks/${name}_chr19.frags.bed
samtools sort -@ 8 -n ${dedup} -o ${nsort}
bedtools bamtobed -bedpe -mate1 -i ${nsort} &gt; ${tags}</code></pre>
<ul>
<li><code>samtools sort -@ 8 -n ${dedup} -o ${nsort}</code>
<ul>
<li>specifying the <code>-n</code> sorts according to read name rather than coordinates by default</li>
</ul></li>
<li><code>bedtools bamtobed -bedpe -mate1 -i ${nsort} &gt; ${tags}</code>
<ul>
<li><code>bedtools bamtobed</code> converts our <code>BAM</code> to <code>BED</code>/coordinates for our reads</li>
<li><code>-bedpe</code> instructs bedtools to report read pairs instead of each read individually</li>
<li>unpaired reads will emit a warning</li>
<li><code>-mate1</code> instructs bedtools to report read1 information first followed by read2</li>
<li>columns : <code>chrR1,startR1,endR1,chrR2,startR2,endR2,readName,mapQ,strandR1,strandR2</code></li>
<li>e.g. <code>chr19 4534039 4534190 chr19 4534110 4534248 SRR20814384.28 60 + -</code></li>
<li>bedtools will check if file is name sorted or coordinate sorted
### Step 3B : Run Peak Caller for ATAC - Tn5 Shift</li>
</ul></li>
<li>perform shift to account for Tn5 binding as a dimer and inserts two adapters separated by 9 bp</li>
</ul>
<div class="callout-blue">
<span class="callout-icon"><i class="fas fa-circle-info" aria-hidden="true"></i></span>
<div class="callout-content">
<div class="callout-title-static">
<p>Note</p>
</div>
<p>Tn5 shift can be skipped if one is not interested in footprinting.</p>
</div>
</div>
<p><strong>Code:</strong></p>
<pre><code>###Shell###
name=MCF10A_ATAC

tags=~/workspace/module123/peaks/${name}_chr19.frags.bed
tn5_tags=~/workspace/module123/peaks/${name}_chr19.frags.tn5.bed


cat ${tags} \
| awk &#39;BEGIN{{OFS=&quot;\t&quot;}}{{printf &quot;%s\t%s\t%s\tN\t1000\t%s\n%s\t%s\t%s\tN\t1000\t%s\n&quot;,$1,$2,$3,$9,$4,$5,$6,$10}}&#39;\
| awk &#39;BEGIN{{OFS = &quot;\t&quot;}} {{if ($6 == &quot;+&quot;) {{$2 = $2 + 4}} else if ($6 == &quot;-&quot;) {{$3 = $3 - 5}} if ($2 &gt;= $3) {{ if ($6 == &quot;+&quot;) {{$2 = $3 - 1}} else {{$3 = $2 + 1}} }} print $0}}&#39;\
&gt; ${tn5_tags}</code></pre>
<ul>
<li><code>cat ${tags} |  awk 'BEGIN{{OFS="\t"}}{{printf "%s\t%s\t%s\tN\t1000\t%s\n%s\t%s\t%s\tN\t1000\t%s\n",$1,$2,$3,$9,$4,$5,$6,$10}}' | awk 'BEGIN {{OFS = "\t"}} {{if ($6 == "+") {{$2 = $2 + 4}} else if ($6 == "-") {{$3 = $3 - 5}} if ($2 &gt;= $3) {{ if ($6 == "+") {{$2 = $3 - 1}} else {{$3 = $2 + 1}} }} print $0}}' &gt; ${tn5_tags}</code>
<ul>
<li><code>read tagFile | rearrange columns | shift Tag depending on strand</code></li>
<li>see below for a more through breakdown of code</li>
</ul>
<pre><code>awk &#39;
BEGIN{{OFS=&quot;\t&quot;}}
{
  {
      printf &quot;%s\t%s\t%s\tN\t1000\t%s\n%s\t%s\t%s\tN\t1000\t%s\n&quot;,
      $1,$2,$3,$9,$4,$5,$6,$10
  }
}&#39;</code></pre></li>
<li><code>BEGIN{{OFS="\t"}}</code> output file as tab delimited</li>
<li><code>printf "%s\t%s\t%s\tN\t1000\t%s\n%s\t%s\t%s\tN\t1000\t%s\n",$1,$2...</code> print string where <code>%s</code> is substituted with the next specified element
<ul>
<li>equivalent to <code>print chrR1,startR1,endR1,"N","1000",strandR1,chrR2,startR2,endR2,strandR2</code></li>
</ul>
<pre><code>awk &#39;
BEGIN {{OFS = &quot;\t&quot;}} 
{
  {
      if ($6 == &quot;+&quot;) 
          {{$2 = $2 + 4}} 
      else if ($6 == &quot;-&quot;) 
          {{$3 = $3 - 5}} 
      if ($2 &gt;= $3) 
      {
          { 
              if ($6 == &quot;+&quot;) 
              {{$2 = $3 - 1}} 
              else 
              {{$3 = $2 + 1}} 
          }
      } print $0
  }
}&#39; 
&gt; ${tn5_tags}</code></pre></li>
<li>if fragment is on <code>+</code> shifted 4 bp to the right</li>
<li>if fragment is on <code>-</code> shift 5 bp to the left
<ul>
<li>while this is recommended, others have shifted <code>4/4</code> instead of <code>4/5</code><br />
</li>
</ul></li>
<li>after the fix is start coordinates is greater than</li>
</ul>
</div>
<div id="step-3c-run-peak-caller-for-atac---peak-calling" class="section level4 hasAnchor">
<h4>Step 3C : Run Peak Caller for ATAC - Peak calling<a href="module-2-alignment-peak-calling-and-visualization.html#step-3c-run-peak-caller-for-atac---peak-calling" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p><strong>Code:</strong></p>
<pre><code>###Shell###
name=MCF10A_ATAC
tn5_tags=~/workspace/module123/peaks/${name}_chr19.frags.tn5.bed

macs2 callpeak \
-t ${tn5_tags} \
-f BED \
-n ${name} \
-g 58617616 \
-p 0.01 \
--shift -100 \
--extsize 200 \
--nomodel \
--bdg \
--keep-dup all \
--outdir ~/workspace/module123/peaks/</code></pre>
<ul>
<li><code>macs2 callpeak -t /home/ubuntu/workspace/module123/peaks/MCF10A_ATAC_chr19.frags.tn5.bed -f BED -n {name} -g 58617616 -p 0.01 --nomodel --extsize 200 --shift -100 --bdg --keep-dup all</code>
<ul>
<li><code>-f BED</code> we specify a <code>BED</code> format input instead of <code>BAM</code></li>
<li><code>-name ${name}</code> name of file</li>
<li><code>-g 58617616</code> size of chr19</li>
<li><code>-p 0.01</code> Pvalue cutoff</li>
<li><code>--nomodel</code> off by default, normally calculates the <code>extsize</code> and <code>shift</code> parameters</li>
<li><code>--extsize 200</code>, b/c Tn5’s activity is on the <code>5'</code>, we extend the <code>5'</code> to get more representation</li>
<li><code>--shift -100</code> should be <code>-1*(extsize/2)</code>, this focuses <code>MACS</code> on our <code>5'</code></li>
<li><code>--bdg</code> generate a pileup bedgraph</li>
<li><code>--keep-dup all</code> retain “duplicates”. As we’ve already filtered out duplicates, MACS2 will call duplicates via genomic coordinates</li>
<li><code>--call-summits</code></li>
</ul></li>
</ul>
</div>
<div id="step-4-blacklist-removal" class="section level4 hasAnchor">
<h4>Step 4 : Blacklist removal<a href="module-2-alignment-peak-calling-and-visualization.html#step-4-blacklist-removal" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<ul>
<li><a href="https://www.nature.com/articles/s41598-019-45839-z">Problematic regions</a> can obscure our results, thus filter any peaks that coincide with those regions.</li>
</ul>
<p><strong>Code:</strong></p>
<pre><code>###Shell###
wget https://www.encodeproject.org/files/ENCFF356LFX/@@download/ENCFF356LFX.bed.gz -O ~/workspace/module123/resources/hg38_blacklist.bed.gz

gunzip ~/workspace/module123/resources/hg38_blacklist.bed.gz

blacklist=~/workspace/module123/resources/hg38_blacklist.bed

sample=&quot;MCF10A_H3K27ac_peaks&quot;
bedtools intersect -v -a ~/workspace/module123/peaks/${sample}.narrowPeak -b ${blacklist} &gt; ~/workspace/module123/peaks/${sample}.blacklistRemoved.narrowPeak
bedtools intersect -u -a ~/workspace/module123/peaks/${sample}.narrowPeak -b ${blacklist} &gt; ~/workspace/module123/peaks/${sample}.blacklist.narrowPeak

sample=&quot;MCF10A_H3K4me3_peaks&quot;
bedtools intersect -v -a ~/workspace/module123/peaks/${sample}.narrowPeak -b ${blacklist} &gt; ~/workspace/module123/peaks/${sample}.blacklistRemoved.narrowPeak
bedtools intersect -u -a ~/workspace/module123/peaks/${sample}.narrowPeak -b ${blacklist} &gt; ~/workspace/module123/peaks/${sample}.blacklist.narrowPeak

sample=&quot;MCF10A_H3K27me3_peaks&quot;
bedtools intersect -v -a ~/workspace/module123/peaks/${sample}.broadPeak -b ${blacklist} &gt; ~/workspace/module123/peaks/${sample}.blacklistRemoved.broadPeak
bedtools intersect -u -a ~/workspace/module123/peaks/${sample}.broadPeak -b ${blacklist} &gt; ~/workspace/module123/peaks/${sample}.blacklist.broadPeak

sample=&quot;MCF10A_ATAC_peaks&quot;
bedtools intersect -v -a ~/workspace/module123/peaks/${sample}.narrowPeak -b ${blacklist} &gt; ~/workspace/module123/peaks/${sample}.blacklistRemoved.narrowPeak
bedtools intersect -u -a ~/workspace/module123/peaks/${sample}.narrowPeak -b ${blacklist} &gt; ~/workspace/module123/peaks/${sample}.blacklist.narrowPeak</code></pre>
<ul>
<li><code>bedtools intersect -u -a ~/workspace/module123/peaks/${sample}.narrowPeak -b ${blacklist}</code>
<ul>
<li><code>bedtools intersect</code> identify elements from <code>fileA</code> and <code>fileB</code> that overlap genomic coordiante wise</li>
<li><code>-u</code> by default, bedtools will output every time a element intersection is detected i.e. if fileA_ele1 overlaps fileB_ele1 and fileB_ele2. The <code>-u</code> instead reports the element once regardless of how many overlaps</li>
</ul></li>
<li><code>bedtools intersect -v -a ~/workspace/module123/peaks/${sample}.narrowPeak -b ${blacklist}</code>
<ul>
<li><code>-v</code> reverse the behaviour, identify elements that do not overlap</li>
</ul></li>
<li>we’ll return to this later when we can visualize the peaks</li>
</ul>
</div>
<div id="step-5a-visualization-of-pileup-tracks" class="section level4 hasAnchor">
<h4>Step 5A : Visualization of pileup tracks<a href="module-2-alignment-peak-calling-and-visualization.html#step-5a-visualization-of-pileup-tracks" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<ul>
<li>in the next steps, we convert our pipleup bedgraphs and bed peak files into a smaller managable formats.</li>
</ul>
<p><strong>Code:</strong></p>
<pre><code>###Shell###
mkdir -p ~/workspace/module123/{bigBed,bigWig}
wget https://hgdownload.cse.ucsc.edu/goldenpath/hg38/bigZips/hg38.chrom.sizes -O workspace/module123/resources/hg38.chrom.sizes

sample=&quot;MCF10A_H3K27ac&quot;
chrom_sizes=~/workspace/module123/resources/hg38.chrom.sizes
input_bedgraph=~/workspace/module123/peaks/${sample}_treat_pileup.bdg
output_bigwig=~/workspace/module123/bigWig/${sample}_treat_pileup.bw

sort -k1,1 -k2,2n ${input_bedgraph} &gt; ~/workspace/module123/bigWig/tmp
bedGraphToBigWig ~/workspace/module123/bigWig/tmp ${chrom_sizes} ${output_bigwig}
rm ~/workspace/module123/bigWig/tmp

input_bedgraph=~/workspace/module123/peaks/MCF10A_H3K27me3_control_lambda.bdg
output_bigwig=~/workspace/module123/bigWig/MCF10A_Input_control_pileup.bw

sort -k1,1 -k2,2n ${input_bedgraph} &gt; ~/workspace/module123/bigWig/tmp
bedGraphToBigWig ~/workspace/module123/bigWig/tmp ${chrom_sizes} ${output_bigwig}
rm ~/workspace/module123/bigWig/tmp</code></pre>
<ul>
<li><code>sort -k1,1 -k2,2n ${input_bedgraph}</code> sorting the BED file
<ul>
<li><code>-k1,1</code> sort first by chromosome alphabetically</li>
<li><code>-k2,2n</code> sort secondarily by genomic coordinates</li>
</ul></li>
<li><code>bedGraphToBigWig</code> convert bedGraph file to bigWig</li>
</ul>
</div>
<div id="step-5b-visualization-of-pileup-tracks-continued" class="section level4 hasAnchor">
<h4>Step 5B : Visualization of pileup tracks continued<a href="module-2-alignment-peak-calling-and-visualization.html#step-5b-visualization-of-pileup-tracks-continued" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<ul>
<li>we’re converting the <code>bedgraph</code> file to a <code>bigWig</code> file.</li>
<li>note <code>bedgraph</code> is an extension of bed(genomic coordinates) + a column with a numeric value
<ul>
<li>in our case pileup/coverage</li>
</ul></li>
<li>a bigWig is a binary version of a <code>wig</code> file
<ul>
<li>wig has a different format : <a href="https://useast.ensembl.org/info/website/upload/wig.html" class="uri">https://useast.ensembl.org/info/website/upload/wig.html</a></li>
<li>the preferred convention for displaying data on a track</li>
</ul></li>
</ul>
<p><strong>Code:</strong></p>
<pre><code>###Shell###
sample=&quot;MCF10A_H3K27me3&quot;
chrom_sizes=~/workspace/module123/resources/hg38.chrom.sizes
input_bedgraph=~/workspace/module123/peaks/${sample}_treat_pileup.bdg
output_bigwig=~/workspace/module123/bigWig/${sample}_treat_pileup.bw

sort -k1,1 -k2,2n ${input_bedgraph} &gt; ~/workspace/module123/bigWig/tmp
bedGraphToBigWig ~/workspace/module123/bigWig/tmp ${chrom_sizes} ${output_bigwig}
rm ~/workspace/module123/bigWig/tmp

sample=&quot;MCF10A_H3K4me3&quot;
chrom_sizes=~/workspace/module123/resources/hg38.chrom.sizes
input_bedgraph=~/workspace/module123/peaks/${sample}_treat_pileup.bdg
output_bigwig=~/workspace/module123/bigWig/${sample}_treat_pileup.bw

sort -k1,1 -k2,2n ${input_bedgraph} &gt; ~/workspace/module123/bigWig/tmp
bedGraphToBigWig ~/workspace/module123/bigWig/tmp ${chrom_sizes} ${output_bigwig}
rm ~/workspace/module123/bigWig/tmp

sample=&quot;MCF10A_ATAC&quot;
chrom_sizes=~/workspace/module123/resources/hg38.chrom.sizes
input_bedgraph=~/workspace/module123/peaks/${sample}_treat_pileup.bdg
output_bigwig=~/workspace/module123/bigWig/${sample}_treat_pileup.bw

sort -k1,1 -k2,2n ${input_bedgraph} &gt; ~/workspace/module123/bigWig/tmp
bedGraphToBigWig ~/workspace/module123/bigWig/tmp ${chrom_sizes} ${output_bigwig}
rm ~/workspace/module123/bigWig/tmp</code></pre>
</div>
<div id="step-5c-visualization-of-peak-tracks" class="section level4 hasAnchor">
<h4>Step 5C : Visualization of peak tracks<a href="module-2-alignment-peak-calling-and-visualization.html#step-5c-visualization-of-peak-tracks" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<ul>
<li>next we convert our <code>BED</code> files</li>
</ul>
<p><strong>Code:</strong></p>
<pre><code>###Shell###
mkdir -p ~/workspace/module123/{bigBed,bigWig}
wget https://hgdownload.cse.ucsc.edu/goldenpath/hg38/bigZips/hg38.chrom.sizes -O workspace/module123/resources/hg38.chrom.sizes

sample=&quot;MCF10A_H3K27ac&quot;
chrom_sizes=~/workspace/module123/resources/hg38.chrom.sizes
input_bed=~/workspace/module123/peaks/${sample}_peaks.blacklistRemoved.narrowPeak
output_bigwig=~/workspace/module123/bigBed/${sample}.blacklistRemoved.bb


sort -k1,1 -k2,2n ${input_bed} | cut -f1-3 &gt; ~/workspace/module123/bigBed/tmp
bedToBigBed ~/workspace/module123/bigBed/tmp ${chrom_sizes} ${output_bigwig}
rm ~/workspace/module123/bigBed/tmp</code></pre>
<ul>
<li><code>sort -k1,1 -k2,2n ${input_bedgraph}</code> sorting the BED file
<ul>
<li><code>-k1,1</code> sort first by chromosome alphabetically</li>
<li><code>-k2,2n</code> sort secondarily by genomic coordinates</li>
</ul></li>
<li><code>bedGraphToBigWig</code> convert bedGraph file to bigBed</li>
</ul>
</div>
<div id="step-5d-visualization-of-peak-tracks-continued" class="section level4 hasAnchor">
<h4>Step 5D : Visualization of peak tracks continued<a href="module-2-alignment-peak-calling-and-visualization.html#step-5d-visualization-of-peak-tracks-continued" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p><strong>Code:</strong></p>
<pre><code>###Shell###
sample=&quot;MCF10A_ATAC&quot;
chrom_sizes=~/workspace/module123/resources/hg38.chrom.sizes
input_bed=~/workspace/module123/peaks/${sample}_peaks.blacklistRemoved.narrowPeak
output_bigwig=~/workspace/module123/bigBed/${sample}.blacklistRemoved.bb


sort -k1,1 -k2,2n ${input_bed} | cut -f1-3 &gt; ~/workspace/module123/bigBed/tmp
bedToBigBed ~/workspace/module123/bigBed/tmp ${chrom_sizes} ${output_bigwig}
rm ~/workspace/module123/bigBed/tmp

sample=&quot;MCF10A_H3K4me3&quot;
chrom_sizes=~/workspace/module123/resources/hg38.chrom.sizes
input_bed=~/workspace/module123/peaks/${sample}_peaks.blacklistRemoved.narrowPeak
output_bigwig=~/workspace/module123/bigBed/${sample}.blacklistRemoved.bb


sort -k1,1 -k2,2n ${input_bed} | cut -f1-3 &gt; ~/workspace/module123/bigBed/tmp
bedToBigBed ~/workspace/module123/bigBed/tmp ${chrom_sizes} ${output_bigwig}
rm ~/workspace/module123/bigBed/tmp

sample=&quot;MCF10A_H3K27me3&quot;
chrom_sizes=~/workspace/module123/resources/hg38.chrom.sizes
input_bed=~/workspace/module123/peaks/${sample}_peaks.blacklistRemoved.broadPeak
output_bigwig=~/workspace/module123/bigBed/${sample}.blacklistRemoved.bb


sort -k1,1 -k2,2n ${input_bed} | cut -f1-3 &gt; ~/workspace/module123/bigBed/tmp
bedToBigBed ~/workspace/module123/bigBed/tmp ${chrom_sizes} ${output_bigwig}
rm ~/workspace/module123/bigBed/tmp</code></pre>
</div>
<div id="step-5e-visualization-of-peaks-and-tracks" class="section level4 hasAnchor">
<h4>Step 5E : Visualization of peaks and tracks<a href="module-2-alignment-peak-calling-and-visualization.html#step-5e-visualization-of-peaks-and-tracks" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<ul>
<li>can either download your tracks and load them from files or via URL</li>
<li>colours used:
<ul>
<li>H3K27ac : blue</li>
<li>H3K27me3 : Brown</li>
<li>H3K4me3 : Green</li>
<li>ATAC : LightBlue</li>
<li>Black : Input</li>
</ul></li>
</ul>
<p><img src="https://github.com/bioinformaticsdotca/EPI_2023/blob/module123/module123_images/BCL3.png?raw=true" alt="Region" width="750" /></p>
<ul>
<li>BCL3 start site is enriched with H3K4me3 and H3K27ac and accessible</li>
</ul>
<p><img src="https://github.com/bioinformaticsdotca/EPI_2023/blob/module123/module123_images/RDH8.png?raw=true" alt="Region" width="750" /></p>
<ul>
<li>Genes around RDH8 show enrichment of H3K27me3 and lack of accessibility and enrichment for H3K27ac and H3K4me3</li>
</ul>
<p><img src="https://github.com/bioinformaticsdotca/EPI_2023/blob/module123/module123_images/blacklist.png?raw=true" alt="Region" width="750" /></p>
<ul>
<li>Blacklist identified region hightlighted in red, represented by both H3K27me3 and ATAC</li>
</ul>
</div>
<div id="step-6a-quality-control-enrichment-in-key-genomic-areas" class="section level4 hasAnchor">
<h4>Step 6A : Quality Control (Enrichment in key genomic areas)<a href="module-2-alignment-peak-calling-and-visualization.html#step-6a-quality-control-enrichment-in-key-genomic-areas" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<ul>
<li>A way to determine the efficacy of your enrichment is to benchmark % of reads to called peaks (FRIP) or known regions</li>
<li>in our toy example we’ll be examining enrichment at promoters (TSS+/- 2kb) and encode defining enhancer regions</li>
<li>For H3K27me3 we’d typically look at HOX regions however no HOX regions on chr19</li>
</ul>
<p><strong>Code :</strong></p>
<pre><code>###Shell###
mkdir workspace/module123/qc
wget https://www.bcgsc.ca/downloads/esu/touchdown/hg38v79_genes_tss_2000.bed -O workspace/module123/resources/hg38v79_genes_tss_2000.bed

sort -k1,1 -k2,2n workspace/module123/resources/hg38v79_genes_tss_2000.bed &gt; tmp
mv tmp workspace/module123/resources/hg38v79_genes_tss_2000.bed

wget https://www.bcgsc.ca/downloads/esu/touchdown/encode_enhancers_liftover.bed -O workspace/module123/resources/encode_enhancers_liftover.bed

TSS=~/workspace/module123/resources/hg38v79_genes_tss_2000.bed
ENH=~/workspace/module123/resources/encode_enhancers_liftover.bed

sample=&quot;MCF10A_H3K27ac&quot;
query_bam=~/workspace/module123/alignments/${sample}_chr19.sorted.dup_marked.bam

samtools view -@4 -q 10 -F 1028 $query_bam -c
samtools view -@4 -q 10 -F 1028 $query_bam -L ${ENH} -c
samtools view -@4 -q 10 -F 1028 $query_bam -L ~/workspace/module123/peaks/${sample}_peaks.blacklistRemoved.narrowPeak -c</code></pre>
<ul>
<li><code>samtools view -@4 -q 10 -F 1028 $query_bam -L ~/workspace/module123/peaks/${sample}_peaks.blacklistRemoved.narrowPeak -c</code>
<ul>
<li><code>samtools view</code> parse through our BAM file</li>
<li><code>-@4</code> resources to use</li>
<li><code>-q 10</code> filter for reads with mapping quality(MAPQ) &gt;10</li>
<li><code>-F 1028</code> Remove reads that have the following flags:UNMAP,DUP</li>
<li><code>$query_bam</code> Bam of interest</li>
<li><code>-c</code> count the number of reads that fulfil the criteria</li>
<li><code>-L</code> perform actions on reads that fall within the specified genomic coordiantes (Bed File)</li>
</ul></li>
</ul>
</div>
<div id="step-6b-quality-control-enrichment-in-key-genomic-areas-continued" class="section level4 hasAnchor">
<h4>Step 6B : Quality Control (Enrichment in key genomic areas) Continued:<a href="module-2-alignment-peak-calling-and-visualization.html#step-6b-quality-control-enrichment-in-key-genomic-areas-continued" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<p><strong>Code:</strong></p>
<pre><code>###Shell###
TSS=~/workspace/module123/resources/hg38v79_genes_tss_2000.bed
ENH=~/workspace/module123/resources/encode_enhancers_liftover.bed

for histone in H3K27ac H3K27me3 H3K4me3 ATAC input;
do
    sample=&quot;MCF10A_${histone}&quot;
    query_bam=~/workspace/module123/alignments/${sample}_chr19.sorted.dup_marked.bam
    samtools view -@4 -q 10 -F 1028 $query_bam -c &gt; ~/workspace/module123/qc/col_${histone}
    samtools view -@4 -q 10 -F 1028 $query_bam -L ${TSS} -c &gt;&gt; ~/workspace/module123/qc/col_${histone}
    samtools view -@4 -q 10 -F 1028 $query_bam -L ${ENH} -c &gt;&gt; ~/workspace/module123/qc/col_${histone}

    if [[ &quot;$histone&quot; == &quot;H3K27me3&quot; ]]; then
        peaks=~/workspace/module123/peaks/${sample}_peaks.blacklistRemoved.broadPeak
        samtools view -@4 -q 10 -F 1028 $query_bam -L ~/workspace/module123/peaks/${sample}_peaks.blacklistRemoved.broadPeak -c &gt;&gt; ~/workspace/module123/qc/col_${histone}
    elif [[ &quot;$histone&quot; == &quot;H3K27ac&quot; || &quot;$histone&quot; == &quot;H3K4me3&quot; || &quot;$histone&quot; == &quot;ATAC&quot; ]]; then
        peaks=~/workspace/module123/peaks/${sample}_peaks.blacklistRemoved.narrowPeak
        samtools view -@4 -q 10 -F 1028 $query_bam -L ~/workspace/module123/peaks/${sample}_peaks.blacklistRemoved.narrowPeak -c &gt;&gt; ~/workspace/module123/qc/col_${histone}
    else 
        echo
    fi
done

paste ~/workspace/module123/qc/col_H3K27ac ~/workspace/module123/qc/col_H3K27me3 ~/workspace/module123/qc/col_H3K4me3 ~/workspace/module123/qc/col_ATAC ~/workspace/module123/qc/col_input &gt; ~/workspace/module123/qc/integers.tsv
paste \
&lt;(awk &#39;{print $1/442829*100}&#39; ~/workspace/module123/qc/col_H3K27ac) \
&lt;(awk &#39;{print $1/1610910*100}&#39; ~/workspace/module123/qc/col_H3K27me3) \
&lt;(awk &#39;{print $1/1512352*100}&#39; ~/workspace/module123/qc/col_H3K4me3) \
&lt;(awk &#39;{print $1/2751833*100}&#39; ~/workspace/module123/qc/col_ATAC) \
&lt;(awk &#39;{print $1/1023265*100}&#39; ~/workspace/module123/qc/col_input) &gt; ~/workspace/module123/qc/percentages.tsv</code></pre>
<p><code>paste ~/workspace/module123/qc/col_H3K27ac ~/workspace/module123/qc/col_H3K27me3 ~/workspace/module123/qc/col_H3K4me3 ~/workspace/module123/qc/col_ATAC ~/workspace/module123/qc/col_input &gt; ~/workspace/module123/qc/integers.tsv</code>
- <code>paste</code> lets us aggregate our results where each is a column)
-
- Reads enrichment in key region</p>
<table style="width:100%;">
<colgroup>
<col width="16%" />
<col width="16%" />
<col width="16%" />
<col width="16%" />
<col width="16%" />
<col width="16%" />
</colgroup>
<thead>
<tr class="header">
<th></th>
<th>H3K27ac</th>
<th>H3K27me3</th>
<th>H3K4me3</th>
<th>ATAC</th>
<th>Input</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>Total</td>
<td>442829</td>
<td>1610910</td>
<td>1512352</td>
<td>2751833</td>
<td>1023265</td>
</tr>
<tr class="even">
<td>TSS</td>
<td>127577</td>
<td>298326</td>
<td>1087032</td>
<td>924326</td>
<td>194996</td>
</tr>
<tr class="odd">
<td>Enhancer</td>
<td>242489</td>
<td>760089</td>
<td>834053</td>
<td>1721794</td>
<td>514115</td>
</tr>
<tr class="even">
<td>In Peaks</td>
<td>69584</td>
<td>783294</td>
<td>1239717</td>
<td>1082674</td>
<td></td>
</tr>
</tbody>
</table>
<ul>
<li>Reads % enrichment in key region</li>
</ul>
<table style="width:100%;">
<colgroup>
<col width="16%" />
<col width="16%" />
<col width="16%" />
<col width="16%" />
<col width="16%" />
<col width="16%" />
</colgroup>
<thead>
<tr class="header">
<th></th>
<th>H3K27ac</th>
<th>H3K27me3</th>
<th>H3K4me3</th>
<th>ATAC</th>
<th>Input</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>TSS</td>
<td>28.8095</td>
<td>18.5191</td>
<td>71.8769</td>
<td>33.5895</td>
<td>19.0563</td>
</tr>
<tr class="even">
<td>Enhancer</td>
<td>54.7591</td>
<td>47.1838</td>
<td>55.1494</td>
<td>62.569</td>
<td>50.2426</td>
</tr>
<tr class="odd">
<td>In Peaks</td>
<td>15.7135</td>
<td>48.6243</td>
<td>81.9728</td>
<td>39.3437</td>
<td></td>
</tr>
</tbody>
</table>
<ul>
<li><code>H3K27ac</code> FRIP is poor as a result of the poorer sequencing depth.</li>
<li><code>H3K4me3</code> is performing well: high FRIP and enriched in promoter regions</li>
</ul>
</div>
</div>
<div id="server-resources-1" class="section level3 hasAnchor">
<h3>Server resources<a href="module-2-alignment-peak-calling-and-visualization.html#server-resources-1" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<div id="qc-resources-1" class="section level4 hasAnchor">
<h4>QC Resources<a href="module-2-alignment-peak-calling-and-visualization.html#qc-resources-1" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<pre><code>####TSS+/-2kb
mkdir workspace/module123/qc
wget https://www.bcgsc.ca/downloads/esu/touchdown/hg38v79_genes_tss_2000.bed -O workspace/module123/resources/hg38v79_genes_tss_2000.bed

sort -k1,1 -k2,2n workspace/module123/resources/hg38v79_genes_tss_2000.bed &gt; tmp
mv tmp workspace/module123/resources/hg38v79_genes_tss_2000.bed

####Enhancer liftover
wget https://www.bcgsc.ca/downloads/esu/touchdown/encode_enhancers_liftover.bed -O workspace/module123/resources/encode_enhancers_liftover.bed

####Blacklist
wget https://www.encodeproject.org/files/ENCFF356LFX/@@download/ENCFF356LFX.bed.gz -O ~/workspace/module123/resources/hg38_blacklist.bed.gz

gunzip ~/workspace/module123/resources/hg38_blacklist.bed.gz</code></pre>
<ul>
<li><code>hg38v79_genes_tss_2000.bed</code>
<ul>
<li>Generated by downloading Ensemblv79 GTF convert to Bed +/-2kb of TSS. See <a href="https://www.biostars.org/p/56280/" class="uri">https://www.biostars.org/p/56280/</a></li>
</ul></li>
<li><code>encode_enhancers_liftover.bed</code>
<ul>
<li>download various <a href="https://egg2.wustl.edu/roadmap/data/byFileType/chromhmmSegmentations/ChmmModels/coreMarks/jointModel/final/">ChroHMM state7</a> and merge</li>
</ul></li>
</ul>
</div>
<div id="encode-bed-1" class="section level4 hasAnchor">
<h4>Encode Bed<a href="module-2-alignment-peak-calling-and-visualization.html#encode-bed-1" class="anchor-section" aria-label="Anchor link to header"></a></h4>
<pre><code>ls ~/CourseData/EPI_data/module123/encode_bed
basal.H3K27ac.peak_calls.bed
basal.H3K27me3.peak_calls.bed
basal.H3K4me1.peak_calls.bed
basal.H3K4me3.peak_calls.bed
lp.H3K27ac.peak_calls.bed
lp.H3K27me3.peak_calls.bed
lp.H3K4me1.peak_calls.bed
lp.H3K4me3.peak_calls.bed
luminal.H3K27ac.peak_calls.bed
luminal.H3K27me3.peak_calls.bed
luminal.H3K4me1.peak_calls.bed
luminal.H3K4me3.peak_calls.bed
stromal.H3K27ac.peak_calls.bed
stromal.H3K27me3.peak_calls.bed
stromal.H3K4me1.peak_calls.bed
stromal.H3K4me3.peak_calls.bed</code></pre>
<ul>
<li><a href="https://epigenomesportal.ca/tracks/CEEHRC/hg38/" class="uri">https://epigenomesportal.ca/tracks/CEEHRC/hg38/</a></li>
<li>Breast Basal CEMT0035</li>
<li>Breast Stromal CEMT0036</li>
<li>Breast Luminal CEMT0037</li>
<li>Breast Luminal Progenitor CEMT0038
#### Encode BigWig</li>
</ul>
<pre><code>ls ~/CourseData/EPI_data/module123/encode_bigWig
basal.H3K27ac.signal_unstranded.bigWig
basal.H3K27me3.signal_unstranded.bigWig
basal.H3K4me1.signal_unstranded.bigWig
basal.H3K4me3.signal_unstranded.bigWig
lp.H3K27ac.signal_unstranded.bigWig
lp.H3K27me3.signal_unstranded.bigWig
lp.H3K4me1.signal_unstranded.bigWig
lp.H3K4me3.signal_unstranded.bigWig
luminal.H3K27ac.signal_unstranded.bigWig
luminal.H3K27me3.signal_unstranded.bigWig
luminal.H3K4me1.signal_unstranded.bigWig
luminal.H3K4me3.signal_unstranded.bigWig
stromal.H3K27ac.signal_unstranded.bigWig
stromal.H3K27me3.signal_unstranded.bigWig
stromal.H3K4me1.signal_unstranded.bigWig
stromal.H3K4me3.signal_unstranded.bigWig</code></pre>
<ul>
<li><a href="https://epigenomesportal.ca/tracks/CEEHRC/hg38/" class="uri">https://epigenomesportal.ca/tracks/CEEHRC/hg38/</a></li>
<li>Breast Basal CEMT0035</li>
<li>Breast Stromal CEMT0036</li>
<li>Breast Luminal CEMT0037</li>
<li>Breast Luminal Progenitor CEMT0038
#### MCF10A Fastq</li>
</ul>
<pre><code>ls ~/CourseData/EPI_data/module123/fastq
MCF10A.ATAC.chr19.R1.fastq.gz
MCF10A.ATAC.chr19.R2.fastq.gz
MCF10A.H3K27ac.chr19.R1.fastq.gz
MCF10A.H3K27ac.chr19.R2.fastq.gz
MCF10A.H3K27me3.chr19.R1.fastq.gz
MCF10A.H3K27me3.chr19.R2.fastq.gz
MCF10A.H3K4me3.chr19.R1.fastq.gz
MCF10A.H3K4me3.chr19.R2.fastq.gz
MCF10A.Input.chr19.R1.fastq.gz
MCF10A.Input.chr19.R2.fastq.gz</code></pre>
<ul>
<li>MCF10A histone marks and input come courtesy of Dr.Hirst</li>
<li>ATACseq data originates from <a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSM6431322">GSM6431322</a>
#### Triplicates</li>
</ul>
<pre><code>CourseData/EPI_data/module123/triplicates/triplicates.csv

CourseData/EPI_data/module123/triplicates/alignments:
MCF10A_H3K4me3_chr19.CondA.Rep1.bam      MCF10A_H3K4me3_chr19.CondB.Rep2.bam      MCF10A_input_chr19.CondA.Rep3.bam
MCF10A_H3K4me3_chr19.CondA.Rep1.bam.bai  MCF10A_H3K4me3_chr19.CondB.Rep2.bam.bai  MCF10A_input_chr19.CondA.Rep3.bam.bai
MCF10A_H3K4me3_chr19.CondA.Rep2.bam      MCF10A_H3K4me3_chr19.CondB.Rep3.bam      MCF10A_input_chr19.CondB.Rep1.bam
MCF10A_H3K4me3_chr19.CondA.Rep2.bam.bai  MCF10A_H3K4me3_chr19.CondB.Rep3.bam.bai  MCF10A_input_chr19.CondB.Rep1.bam.bai
MCF10A_H3K4me3_chr19.CondA.Rep3.bam      MCF10A_input_chr19.CondA.Rep1.bam        MCF10A_input_chr19.CondB.Rep2.bam
MCF10A_H3K4me3_chr19.CondA.Rep3.bam.bai  MCF10A_input_chr19.CondA.Rep1.bam.bai    MCF10A_input_chr19.CondB.Rep2.bam.bai
MCF10A_H3K4me3_chr19.CondB.Rep1.bam      MCF10A_input_chr19.CondA.Rep2.bam        MCF10A_input_chr19.CondB.Rep3.bam
MCF10A_H3K4me3_chr19.CondB.Rep1.bam.bai  MCF10A_input_chr19.CondA.Rep2.bam.bai    MCF10A_input_chr19.CondB.Rep3.bam.bai

CourseData/EPI_data/module123/triplicates/bigWig:
CondA.Rep1.bw  CondA.Rep2.bw  CondA.Rep3.bw  CondB.Rep1.bw  CondB.Rep2.bw  CondB.Rep3.bw

CourseData/EPI_data/module123/triplicates/peaks:
CondA.Rep1_peaks.narrowPeak  CondA.Rep3_peaks.narrowPeak  CondB.Rep2_peaks.narrowPeak
CondA.Rep2_peaks.narrowPeak  CondB.Rep1_peaks.narrowPeak  CondB.Rep3_peaks.narrowPeak</code></pre>
<ul>
<li>triplicates were generated from MCF10A_H3K4me3 by choosing a list of exclusive peaks for condA and condB and randomly subsampling replicates accordingly</li>
</ul>
<div class="callout-green">
<span class="callout-icon"><i class="fas fa-circle-check" aria-hidden="true"></i></span>
<div class="callout-content">
<div class="callout-title-static">
<p>Lab Completed!</p>
</div>
<p>Congratulations! You have completed Lab 2!</p>
</div>
</div>

</div>
</div>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="module-1-introduction-to-chip-sequencing-and-analysis.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="module-3-chip-seq-differential-analysis.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
  "sharing": {
    "github": false,
    "facebook": false,
    "twitter": false,
    "linkedin": false,
    "weibo": false,
    "instapaper": false,
    "vk": false,
    "whatsapp": false,
    "all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
  },
  "fontsettings": {
    "theme": "white",
    "family": "sans",
    "size": 2
  },
  "edit": {
    "link": "https://github.com/bioinformaticsdotca/EPI_2023/blob/main/020-module-2.Rmd",
    "text": "Edit"
  },
  "history": {
    "link": null,
    "text": null
  },
  "view": {
    "link": null,
    "text": null
  },
  "download": ["_main.pdf", "_main.epub"],
  "search": {
    "engine": "fuse",
    "options": null
  },
  "toc": {
    "collapse": "subsection"
  }
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.9/latest.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
